// @neon/env/dist/lib/cli/commands.js
import { fetchEnv, toEntries } from "../env.js";
import { resolveContext } from "./resolve-context.js";
import { existsSync, readFileSync } from "node:fs";
import { spawn } from "node:child_process";
import { dirname, join } from "node:path";
import { ConfigLoadError, ErrorCode, MissingContextError, PlatformError, loadConfigFromFile } from "@neon/config/v1";
//#region src/lib/cli/commands.ts
/** File `env run` reads to layer one-time auth keys. Matches the Vercel/Next.js convention. */
const DEFAULT_ENV_FILE = ".env.local";
/**
* Implementation of `neon-env run -- <cmd...>`. Loads `neon.ts`, fetches the env from
* Neon, then spawns the user-supplied command with the env vars injected on top of the
* inherited `process.env`. Stdio is inherited so interactive dev servers keep working.
* The parent process exits with the child's exit code.
*/
async function runEnvRun(options, ctx) {
	if (options.command.length === 0) return failure([
		"`env run` requires a command to spawn.",
		"Usage: neon-env run -- <command> [args...]",
		"Example: neon-env run -- npm run dev"
	].join("\n"));
	const resolved = resolveContext({
		cwd: ctx.cwd,
		...options.projectId ? { projectId: options.projectId } : {},
		...options.branch ? { branch: options.branch } : {}
	});
	if (!resolved.ok) return failure(["`env run` could not resolve the Neon project and branch:", ...resolved.missing.map((m) => `  - ${m}`)].join("\n"), 3);
	let injected;
	try {
		injected = toEntries(await loadConfigAndFetchEnv(options, ctx, resolved.context));
	} catch (err) {
		return handleError(err);
	}
	const [executable, ...args] = options.command;
	return {
		exitCode: await spawnAndWait(executable, args, {
			cwd: ctx.cwd,
			env: {
				...process.env,
				...injected
			}
		}),
		stdout: "",
		stderr: ""
	};
}
/**
* Implementation of `neon-env export`. Resolves the branch's Neon env the same way `run`
* does (neon.ts policy + linked branch), then writes it to stdout — as dotenv lines or JSON —
* instead of spawning a process, so other env tools can consume it. For example, varlock can
* bulk-load it with `@setValuesBulk(exec("neon-env export --format json"), format=json)`.
*/
async function runEnvExport(options, ctx) {
	const resolved = resolveContext({
		cwd: ctx.cwd,
		...options.projectId ? { projectId: options.projectId } : {},
		...options.branch ? { branch: options.branch } : {}
	});
	if (!resolved.ok) return failure(["`env export` could not resolve the Neon project and branch:", ...resolved.missing.map((m) => `  - ${m}`)].join("\n"), 3);
	let entries;
	try {
		entries = toEntries(await loadConfigAndFetchEnv(options, ctx, resolved.context));
	} catch (err) {
		return handleError(err);
	}
	return {
		exitCode: 0,
		stdout: options.format === "json" ? `${JSON.stringify(entries, null, 2)}\n` : toDotenv(entries),
		stderr: ""
	};
}
/** Render an env map as dotenv `KEY=value` lines, quoting values that need it. */
function toDotenv(entries) {
	const lines = Object.entries(entries).map(([key, value]) => formatDotenvLine(key, value));
	return lines.length > 0 ? `${lines.join("\n")}\n` : "";
}
/**
* Render a single `KEY=value` dotenv line, double-quoting (and escaping) values that contain
* whitespace, `#`, quotes, or `=` so connection strings round-trip through dotenv parsers.
*/
function formatDotenvLine(key, value) {
	if (!/[\s#"'=]/.test(value)) return `${key}=${value}`;
	return `${key}="${value.replace(/\\/g, "\\\\").replace(/"/g, "\\\"")}"`;
}
/**
* Load `neon.ts`, then call `fetchEnv` with the explicitly-resolved project + branch.
* Layers any one-time Auth keys from `.env.local` (next to the config file) into the env
* source so re-runs keep round-tripping values the Neon API only returns once at
* integration-creation time.
*/
async function loadConfigAndFetchEnv(options, ctx, resolved) {
	const { config, resolvedPath } = await loadConfigFromFile({
		...options.configPath ? { path: options.configPath } : {},
		cwd: ctx.cwd
	});
	const envFileSource = join(dirname(resolvedPath), DEFAULT_ENV_FILE);
	const fileEnv = existsSync(envFileSource) ? parseEnvFile(readFileSync(envFileSource, "utf-8")) : {};
	return fetchEnv(config, {
		projectId: resolved.projectId,
		branch: resolved.branch,
		env: {
			...process.env,
			...fileEnv
		},
		...ctx.api ? { api: ctx.api } : {},
		...options.apiKey ? { apiKey: options.apiKey } : {}
	});
}
/**
* Spawn a child process with stdio inherited so dev servers stay interactive. Resolves
* with the child's exit code (treating signal terminations as code 1 so the CLI surfaces
* a non-zero exit consistently).
*/
function spawnAndWait(command, args, options) {
	return new Promise((resolve) => {
		const child = spawn(command, args, {
			cwd: options.cwd,
			env: options.env,
			stdio: "inherit"
		});
		child.on("error", (err) => {
			process.stderr.write(`neon-env run: failed to spawn '${command}': ${err.message}\n`);
			resolve(1);
		});
		child.on("exit", (code, signal) => {
			if (typeof code === "number") {
				resolve(code);
				return;
			}
			if (signal) {
				process.stderr.write(`neon-env run: child terminated by signal ${signal}\n`);
				resolve(1);
				return;
			}
			resolve(1);
		});
	});
}
function parseEnvFile(body) {
	const out = {};
	for (const line of body.split("\n")) {
		const parsed = parseEnvLine(line);
		if (parsed) out[parsed.key] = parsed.value;
	}
	return out;
}
function parseEnvLine(line) {
	const match = line.match(/^\s*(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.*)$/);
	const key = match?.[1];
	const rawValue = match?.[2];
	if (key === void 0 || rawValue === void 0) return null;
	return {
		key,
		value: unescapeEnvValue(rawValue.trim())
	};
}
function unescapeEnvValue(value) {
	if (value.length >= 2 && value.startsWith("\"") && value.endsWith("\"")) return value.slice(1, -1).replace(/\\"/g, "\"").replace(/\\\\/g, "\\");
	if (value.length >= 2 && value.startsWith("'") && value.endsWith("'")) return value.slice(1, -1);
	return value;
}
/**
* Stable exit code per `PlatformError` code. Mirrors the table in the config package so
* shell pipelines can branch on the specific failure mode without parsing free text.
*/
const EXIT_CODE_BY_PLATFORM_ERROR_CODE = {
	[ErrorCode.MissingApiKey]: 1,
	[ErrorCode.Unauthorized]: 6,
	[ErrorCode.Forbidden]: 7,
	[ErrorCode.NotFound]: 8,
	[ErrorCode.RateLimited]: 9,
	[ErrorCode.NetworkError]: 10,
	[ErrorCode.ServerError]: 11,
	[ErrorCode.Locked]: 11,
	[ErrorCode.InternalError]: 99
};
function handleError(err) {
	if (err instanceof MissingContextError) return errorResult(err, `Missing context: ${err.message}`, 3);
	if (err instanceof ConfigLoadError) return errorResult(err, `Failed to load config: ${err.message}`, 4);
	if (err instanceof PlatformError) {
		const exitCode = EXIT_CODE_BY_PLATFORM_ERROR_CODE[err.code];
		if (exitCode !== void 0) return errorResult(err, err.message, exitCode);
		return errorResult(err, `[${err.code}] ${err.message}`, 5);
	}
	if (err instanceof Error) return errorResult(err, err.message, 1);
	return failure(String(err), 1);
}
function errorResult(err, message, exitCode) {
	const result = {
		exitCode,
		stdout: "",
		stderr: `${message}\n`
	};
	const debug = buildDebugInfo(err);
	if (debug) result.debugInfo = debug;
	return result;
}
function buildDebugInfo(err) {
	if (!(err instanceof Error)) return void 0;
	const lines = [];
	if (err instanceof PlatformError) {
		lines.push(`code     : ${err.code}`);
		if (Object.keys(err.details).length > 0) lines.push(`details  : ${JSON.stringify(err.details, null, 2)}`);
	}
	if (err.cause instanceof Error) lines.push(`cause    : ${err.cause.name}: ${err.cause.message}`);
	if (err.stack) lines.push(err.stack);
	return lines.length > 0 ? lines.join("\n") : void 0;
}
function failure(message, exitCode = 1) {
	return {
		exitCode,
		stdout: "",
		stderr: `${message}\n`
	};
}
//#endregion
export { runEnvExport, runEnvRun };
