// @neon/env/dist/lib/cli/resolve-context.js
import { existsSync, readFileSync, statSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { homedir } from "node:os";
//#region src/lib/cli/resolve-context.ts
/**
* Resolve `projectId` and `branch` for a CLI invocation. Precedence (each wins over the
* next): explicit flag → `NEON_*` env var → `.neon[/project.json]` walked up from `cwd`.
*
* Returns the resolved values plus a list of human-readable reasons for any field that
* could not be resolved (so the caller can render one combined error).
*/
function resolveContext(options) {
	const env = options.env ?? process.env;
	const file = findNeonFile(options.cwd);
	const projectId = nonEmpty(options.projectId) ?? nonEmpty(env.NEON_PROJECT_ID) ?? file?.projectId;
	const branch = nonEmpty(options.branch) ?? nonEmpty(env.NEON_BRANCH) ?? nonEmpty(env.NEON_BRANCH_ID) ?? file?.branch;
	const missing = [];
	if (!projectId) missing.push("project id — pass `--project-id`, set `NEON_PROJECT_ID`, or add `projectId` to `.neon` (run `npx neonctl link`).");
	if (!branch) missing.push("branch — pass `--branch`, set `NEON_BRANCH`/`NEON_BRANCH_ID`, or add `branch` to `.neon` (run `npx neonctl link` / `neonctl checkout <branch>`).");
	if (!projectId || !branch) return {
		ok: false,
		missing
	};
	return {
		ok: true,
		context: {
			projectId,
			branch
		}
	};
}
/**
* Walk up from `cwd` looking for `.neon/project.json` (preferred) or `.neon` (neonctl
* convention). Stops at the first `.git` directory or the home directory. Read-only.
*/
function findNeonFile(cwd) {
	let current = resolve(cwd);
	const stop = resolve(homedir());
	let lastSeen = null;
	while (true) {
		const parsed = readNeonFileAt(resolve(current, ".neon", "project.json")) ?? readNeonFileAt(resolve(current, ".neon"));
		if (parsed) return parsed;
		if (current === stop) return null;
		if (existsSync(resolve(current, ".git"))) return null;
		const parent = dirname(current);
		if (parent === current || parent === lastSeen) return null;
		lastSeen = current;
		current = parent;
	}
}
function readNeonFileAt(path) {
	if (!isFile(path)) return null;
	let raw;
	try {
		raw = readFileSync(path, "utf-8");
	} catch {
		return null;
	}
	let parsed;
	try {
		parsed = JSON.parse(raw);
	} catch {
		return null;
	}
	if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) return null;
	const obj = parsed;
	const out = {};
	if (typeof obj.projectId === "string" && obj.projectId !== "") out.projectId = obj.projectId;
	const branch = typeof obj.branch === "string" && obj.branch !== "" ? obj.branch : typeof obj.branchId === "string" && obj.branchId !== "" ? obj.branchId : void 0;
	if (branch) out.branch = branch;
	return out;
}
function isFile(path) {
	try {
		return statSync(path).isFile();
	} catch {
		return false;
	}
}
function nonEmpty(value) {
	if (typeof value !== "string") return void 0;
	const trimmed = value.trim();
	return trimmed === "" ? void 0 : trimmed;
}
//#endregion
export { resolveContext };
