// @neon/sdk/dist/neon/resources/account.js
import { createApiKey, getActiveRegions, getCurrentUserInfo, getCurrentUserOrganizations, listApiKeys, revokeApiKey } from "../../client/sdk.gen.js";
//#region src/neon/resources/account.ts
/** Current user / account resource. */
var User = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	me(opts) {
		return this.#ctx.run(opts, (client) => getCurrentUserInfo({
			client,
			throwOnError: false
		}), (data) => data);
	}
	organizations(opts) {
		return this.#ctx.run(opts, (client) => getCurrentUserOrganizations({
			client,
			throwOnError: false
		}), (data) => data.organizations);
	}
};
/** Active regions. */
var Regions = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	list(opts) {
		return this.#ctx.run(opts, (client) => getActiveRegions({
			client,
			throwOnError: false
		}), (data) => data.regions);
	}
};
/** Account API keys. */
var ApiKeys = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	list(opts) {
		return this.#ctx.run(opts, (client) => listApiKeys({
			client,
			throwOnError: false
		}), (data) => data);
	}
	create(keyName, opts) {
		return this.#ctx.run(opts, (client) => createApiKey({
			client,
			body: { key_name: keyName },
			throwOnError: false
		}), (data) => data);
	}
	revoke(keyId, opts) {
		return this.#ctx.run(opts, (client) => revokeApiKey({
			client,
			path: { key_id: keyId },
			throwOnError: false
		}), (data) => data);
	}
};
//#endregion
export { ApiKeys, Regions, User };
