// @neon/sdk/dist/neon/resources/branches.js
import { NeonError } from "../errors.js";
import { err, finalize, ok } from "../result.js";
import { createProjectBranch, deleteProjectBranch, finalizeRestoreBranch, getProjectBranch, listProjectBranches, setDefaultProjectBranch, updateProjectBranch } from "../../client/sdk.gen.js";
import { withConnectionString } from "../connection.js";
import { paginate } from "../paginate.js";
//#region src/neon/resources/branches.ts
/** Branch resource — one API call per CRUD method, plus the `createWithCompute` workflow. */
var Branches = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	/** @apiCall GET /projects/{project_id}/branches (cursor-paginated) */
	list(projectId, query) {
		return paginate((cursor, signal) => listProjectBranches({
			client: this.#ctx.client,
			path: { project_id: projectId },
			query: {
				...query,
				cursor
			},
			throwOnError: false,
			signal
		}), (data) => ({
			items: data?.branches ?? [],
			cursor: data?.pagination?.next
		}));
	}
	get(projectId, branchId, opts) {
		return this.#ctx.run(opts, (client) => getProjectBranch({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			throwOnError: false
		}), (data) => data.branch);
	}
	create(projectId, input, opts) {
		return this.#ctx.run(opts, (client) => createProjectBranch({
			client,
			path: { project_id: projectId },
			body: { branch: input },
			throwOnError: false
		}), (data) => data.branch);
	}
	update(projectId, branchId, input, opts) {
		return this.#ctx.run(opts, (client) => updateProjectBranch({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			body: { branch: input },
			throwOnError: false
		}), (data) => data.branch);
	}
	delete(projectId, branchId, opts) {
		return this.#ctx.runVoid(opts, (client) => deleteProjectBranch({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			throwOnError: false
		}));
	}
	async createWithCompute(projectId, input, opts) {
		const shouldThrow = opts?.throwOnError ?? this.#ctx.defaults.throwOnError;
		return finalize(withConnectionString(await this.#ctx.execute({
			...opts,
			waitForReadiness: opts?.waitForReadiness ?? true
		}, (client) => createProjectBranch({
			client,
			path: { project_id: projectId },
			body: {
				branch: {
					name: input.name,
					parent_id: input.parentId
				},
				endpoints: [{
					type: "read_write",
					autoscaling_limit_min_cu: input.compute?.minCu,
					autoscaling_limit_max_cu: input.compute?.maxCu,
					suspend_timeout_seconds: input.compute?.suspendTimeoutSeconds
				}]
			},
			throwOnError: false
		}), (data) => data), (data) => data.connection_uris, (data, connectionString) => ({
			branch: data.branch,
			endpoint: data.endpoints[0],
			connectionString
		}), opts?.pooled ?? true), shouldThrow);
	}
	async getDefault(projectId, opts) {
		const shouldThrow = opts?.throwOnError ?? this.#ctx.defaults.throwOnError;
		const result = await this.#ctx.execute(opts, (client) => listProjectBranches({
			client,
			path: { project_id: projectId },
			throwOnError: false
		}), (data) => data.branches);
		if (result.error) return finalize(err(result.error), shouldThrow);
		const branch = result.data.find((candidate) => candidate.default);
		if (!branch) return finalize(err(new NeonError("No default branch found for the project.", "client")), shouldThrow);
		return finalize(ok(branch), shouldThrow);
	}
	setDefault(projectId, branchId, opts) {
		return this.#ctx.run(opts, (client) => setDefaultProjectBranch({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			throwOnError: false
		}), (data) => data.branch);
	}
	finalizeRestore(projectId, branchId, input, opts) {
		return this.#ctx.run(opts, (client) => finalizeRestoreBranch({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			body: { name: input?.name },
			throwOnError: false
		}), () => void 0);
	}
};
//#endregion
export { Branches };
