// @neon/sdk/dist/neon/resources/consumption.js
import { getConsumptionHistoryPerBranchV2, getConsumptionHistoryPerProject, getConsumptionHistoryPerProjectV2 } from "../../client/sdk.gen.js";
import { paginate } from "../paginate.js";
//#region src/neon/resources/consumption.ts
/** Consumption history (cursor-paginated). */
var Consumption = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	/** @apiCall GET /consumption_history/projects */
	perProject(query) {
		return paginate((cursor, signal) => getConsumptionHistoryPerProject({
			client: this.#ctx.client,
			query: {
				...query,
				cursor
			},
			throwOnError: false,
			signal
		}), (data) => ({
			items: data?.projects ?? [],
			cursor: data?.pagination?.cursor
		}));
	}
	/** @apiCall GET /consumption_history/v2/projects */
	perProjectV2(query) {
		return paginate((cursor, signal) => getConsumptionHistoryPerProjectV2({
			client: this.#ctx.client,
			query: {
				...query,
				cursor
			},
			throwOnError: false,
			signal
		}), (data) => ({
			items: data?.projects ?? [],
			cursor: data?.pagination?.cursor
		}));
	}
	/** @apiCall GET /consumption_history/v2/branches */
	perBranchV2(query) {
		return paginate((cursor, signal) => getConsumptionHistoryPerBranchV2({
			client: this.#ctx.client,
			query: {
				...query,
				cursor
			},
			throwOnError: false,
			signal
		}), (data) => ({
			items: data?.branches ?? [],
			cursor: data?.pagination?.cursor
		}));
	}
};
//#endregion
export { Consumption };
