// @neon/sdk/dist/neon/resources/dataapi.js
import { createProjectBranchDataApi, deleteProjectBranchDataApi, getProjectBranchDataApi, updateProjectBranchDataApi } from "../../client/sdk.gen.js";
//#region src/neon/resources/dataapi.ts
/** Neon Data API resource (branch + database scoped). */
var DataApi = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	get(projectId, branchId, databaseName, opts) {
		return this.#ctx.run(opts, (client) => getProjectBranchDataApi({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				database_name: databaseName
			},
			throwOnError: false
		}), (data) => data);
	}
	create(projectId, branchId, databaseName, input, opts) {
		return this.#ctx.run(opts, (client) => createProjectBranchDataApi({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				database_name: databaseName
			},
			body: input,
			throwOnError: false
		}), (data) => data);
	}
	update(projectId, branchId, databaseName, input, opts) {
		return this.#ctx.run(opts, (client) => updateProjectBranchDataApi({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				database_name: databaseName
			},
			body: input,
			throwOnError: false
		}), () => void 0);
	}
	delete(projectId, branchId, databaseName, opts) {
		return this.#ctx.run(opts, (client) => deleteProjectBranchDataApi({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				database_name: databaseName
			},
			throwOnError: false
		}), () => void 0);
	}
};
//#endregion
export { DataApi };
