// @neon/sdk/dist/neon/resources/databases.js
import { createProjectBranchDatabase, deleteProjectBranchDatabase, getProjectBranchDatabase, listProjectBranchDatabases, updateProjectBranchDatabase } from "../../client/sdk.gen.js";
//#region src/neon/resources/databases.ts
/** Database resource (branch-scoped). */
var Databases = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	list(projectId, branchId, opts) {
		return this.#ctx.run(opts, (client) => listProjectBranchDatabases({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			throwOnError: false
		}), (data) => data.databases);
	}
	get(projectId, branchId, name, opts) {
		return this.#ctx.run(opts, (client) => getProjectBranchDatabase({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				database_name: name
			},
			throwOnError: false
		}), (data) => data.database);
	}
	create(projectId, branchId, input, opts) {
		return this.#ctx.run(opts, (client) => createProjectBranchDatabase({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			body: { database: input },
			throwOnError: false
		}), (data) => data.database);
	}
	update(projectId, branchId, name, input, opts) {
		return this.#ctx.run(opts, (client) => updateProjectBranchDatabase({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				database_name: name
			},
			body: { database: input },
			throwOnError: false
		}), (data) => data.database);
	}
	delete(projectId, branchId, name, opts) {
		return this.#ctx.runVoid(opts, (client) => deleteProjectBranchDatabase({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				database_name: name
			},
			throwOnError: false
		}));
	}
};
//#endregion
export { Databases };
