// @neon/sdk/dist/neon/resources/endpoints.js
import { createProjectEndpoint, deleteProjectEndpoint, getProjectEndpoint, listProjectEndpoints, restartProjectEndpoint, startProjectEndpoint, suspendProjectEndpoint, updateProjectEndpoint } from "../../client/sdk.gen.js";
//#region src/neon/resources/endpoints.ts
/** Compute endpoint resource (project-scoped). */
var Endpoints = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	list(projectId, opts) {
		return this.#ctx.run(opts, (client) => listProjectEndpoints({
			client,
			path: { project_id: projectId },
			throwOnError: false
		}), (data) => data.endpoints);
	}
	get(projectId, endpointId, opts) {
		return this.#ctx.run(opts, (client) => getProjectEndpoint({
			client,
			path: {
				project_id: projectId,
				endpoint_id: endpointId
			},
			throwOnError: false
		}), (data) => data.endpoint);
	}
	create(projectId, input, opts) {
		return this.#ctx.run(opts, (client) => createProjectEndpoint({
			client,
			path: { project_id: projectId },
			body: { endpoint: input },
			throwOnError: false
		}), (data) => data.endpoint);
	}
	update(projectId, endpointId, input, opts) {
		return this.#ctx.run(opts, (client) => updateProjectEndpoint({
			client,
			path: {
				project_id: projectId,
				endpoint_id: endpointId
			},
			body: { endpoint: input },
			throwOnError: false
		}), (data) => data.endpoint);
	}
	delete(projectId, endpointId, opts) {
		return this.#ctx.runVoid(opts, (client) => deleteProjectEndpoint({
			client,
			path: {
				project_id: projectId,
				endpoint_id: endpointId
			},
			throwOnError: false
		}));
	}
	start(projectId, endpointId, opts) {
		return this.#ctx.run(opts, (client) => startProjectEndpoint({
			client,
			path: {
				project_id: projectId,
				endpoint_id: endpointId
			},
			throwOnError: false
		}), (data) => data.endpoint);
	}
	suspend(projectId, endpointId, opts) {
		return this.#ctx.run(opts, (client) => suspendProjectEndpoint({
			client,
			path: {
				project_id: projectId,
				endpoint_id: endpointId
			},
			throwOnError: false
		}), (data) => data.endpoint);
	}
	restart(projectId, endpointId, opts) {
		return this.#ctx.run(opts, (client) => restartProjectEndpoint({
			client,
			path: {
				project_id: projectId,
				endpoint_id: endpointId
			},
			throwOnError: false
		}), (data) => data.endpoint);
	}
};
//#endregion
export { Endpoints };
