// @neon/sdk/dist/neon/resources/operations.js
import { getProjectOperation, listProjectOperations } from "../../client/sdk.gen.js";
import { waitForOperations } from "../wait.js";
import { paginate } from "../paginate.js";
//#region src/neon/resources/operations.ts
/** Operation resource — read operations and wait for them to finish. */
var Operations = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	/** @apiCall GET /projects/{project_id}/operations (cursor-paginated) */
	list(projectId) {
		return paginate((cursor, signal) => listProjectOperations({
			client: this.#ctx.client,
			path: { project_id: projectId },
			query: { cursor },
			throwOnError: false,
			signal
		}), (data) => ({
			items: data?.operations ?? [],
			cursor: data?.pagination?.cursor
		}));
	}
	get(projectId, operationId, opts) {
		return this.#ctx.run(opts, (client) => getProjectOperation({
			client,
			path: {
				project_id: projectId,
				operation_id: operationId
			},
			throwOnError: false
		}), (data) => data.operation);
	}
	async waitFor(operations, opts) {
		const defaults = this.#ctx.defaults;
		const shouldThrow = opts?.throwOnError ?? defaults.throwOnError;
		const result = await waitForOperations(this.#ctx.client, operations, {
			pollIntervalMs: opts?.pollIntervalMs ?? defaults.waitOptions.pollIntervalMs,
			timeoutMs: opts?.timeoutMs ?? defaults.waitOptions.timeoutMs,
			signal: opts?.signal ?? defaults.waitOptions.signal
		});
		if (!shouldThrow) return result;
		if (result.error) throw result.error;
		return result.data;
	}
};
//#endregion
export { Operations };
