// @neon/sdk/dist/neon/resources/postgres.js
import { NeonError, toNeonError } from "../errors.js";
import { err, finalize, ok } from "../result.js";
import { getConnectionUri, listProjectBranchDatabases, listProjectBranchRoles, listProjectBranches } from "../../client/sdk.gen.js";
import { DataApi } from "./dataapi.js";
import { Databases } from "./databases.js";
import { Endpoints } from "./endpoints.js";
import { Roles } from "./roles.js";
//#region src/neon/resources/postgres.ts
/**
* The Postgres data plane of a branch: compute endpoints, roles, databases, the Data API,
* and a connection-string helper. (Grouped under `neon.postgres.*` so future top-level
* namespaces like `functions` / `storage` stay unambiguous.)
*/
var Postgres = class {
	endpoints;
	roles;
	databases;
	dataApi;
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
		this.endpoints = new Endpoints(ctx);
		this.roles = new Roles(ctx);
		this.databases = new Databases(ctx);
		this.dataApi = new DataApi(ctx);
	}
	async connectionString(params, opts) {
		const shouldThrow = opts?.throwOnError ?? this.#ctx.defaults.throwOnError;
		return finalize(await this.#resolve(params, opts?.signal), shouldThrow);
	}
	async #resolve(params, signal) {
		const client = this.#ctx.client;
		const projectId = params.projectId;
		let branchId = params.branchId;
		let roleName = params.roleName;
		let databaseName = params.databaseName;
		if ((!roleName || !databaseName) && !branchId) {
			const branches = await listProjectBranches({
				client,
				path: { project_id: projectId },
				throwOnError: false,
				signal
			});
			if (branches.error || !branches.data) return err(toNeonError(branches.error, branches.response));
			branchId = branches.data.branches.find((branch) => branch.default)?.id;
			if (!branchId) return err(new NeonError("Could not determine the default branch; pass branchId.", "client"));
		}
		if (!roleName) {
			if (!branchId) return err(branchRequired("roleName"));
			const roles = await listProjectBranchRoles({
				client,
				path: {
					project_id: projectId,
					branch_id: branchId
				},
				throwOnError: false,
				signal
			});
			if (roles.error || !roles.data) return err(toNeonError(roles.error, roles.response));
			if (roles.data.roles.length !== 1) return err(ambiguous("role", roles.data.roles.length, "roleName"));
			roleName = roles.data.roles[0].name;
		}
		if (!databaseName) {
			if (!branchId) return err(branchRequired("databaseName"));
			const databases = await listProjectBranchDatabases({
				client,
				path: {
					project_id: projectId,
					branch_id: branchId
				},
				throwOnError: false,
				signal
			});
			if (databases.error || !databases.data) return err(toNeonError(databases.error, databases.response));
			if (databases.data.databases.length !== 1) return err(ambiguous("database", databases.data.databases.length, "databaseName"));
			databaseName = databases.data.databases[0].name;
		}
		const conn = await getConnectionUri({
			client,
			path: { project_id: projectId },
			query: {
				branch_id: branchId,
				endpoint_id: params.endpointId,
				database_name: databaseName,
				role_name: roleName,
				pooled: params.pooled ?? true
			},
			throwOnError: false,
			signal
		});
		if (conn.error || !conn.data) return err(toNeonError(conn.error, conn.response));
		return ok(conn.data.uri);
	}
};
function branchRequired(param) {
	return new NeonError(`Pass branchId or ${param} to resolve a connection string.`, "client");
}
function ambiguous(kind, count, param) {
	return new NeonError(`Expected exactly one ${kind} to auto-select for the connection string; found ${count}. Pass ${param}.`, "client");
}
//#endregion
export { Postgres };
