// @neon/sdk/dist/neon/resources/projects.js
import { NeonError } from "../errors.js";
import { err, finalize } from "../result.js";
import { createProject, deleteProject, getProject, listProjects, transferProjectsFromOrgToOrg, transferProjectsFromUserToOrg, updateProject } from "../../client/sdk.gen.js";
import { withConnectionString } from "../connection.js";
import { paginate } from "../paginate.js";
//#region src/neon/resources/projects.ts
/** Project resource — one API call per method (`list` is cursor-paginated). */
var Projects = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	/**
	* List projects (cursor-paginated). Returns a lazy list — `await .all()`, `.page()`,
	* or `for await (… of …)`.
	*
	* @apiCall GET /projects
	*/
	list(query) {
		return paginate((cursor, signal) => listProjects({
			client: this.#ctx.client,
			query: {
				org_id: this.#ctx.defaults.orgId,
				...query,
				cursor
			},
			throwOnError: false,
			signal
		}), (data) => ({
			items: data?.projects ?? [],
			cursor: data?.pagination?.cursor
		}));
	}
	get(id, opts) {
		return this.#ctx.run(opts, (client) => getProject({
			client,
			path: { project_id: id },
			throwOnError: false
		}), (data) => data.project);
	}
	create(input, opts) {
		return this.#ctx.run(opts, (client) => createProject({
			client,
			body: { project: {
				...this.#ctx.defaults.orgId ? { org_id: this.#ctx.defaults.orgId } : {},
				...input
			} },
			throwOnError: false
		}), (data) => data.project);
	}
	async createAndConnect(input, opts) {
		const shouldThrow = opts?.throwOnError ?? this.#ctx.defaults.throwOnError;
		return finalize(withConnectionString(await this.#ctx.execute({
			...opts,
			waitForReadiness: opts?.waitForReadiness ?? true
		}, (client) => createProject({
			client,
			body: { project: {
				...this.#ctx.defaults.orgId ? { org_id: this.#ctx.defaults.orgId } : {},
				...input
			} },
			throwOnError: false
		}), (data) => data), (data) => data.connection_uris, (data, connectionString) => ({
			project: data.project,
			connectionString
		}), opts?.pooled ?? true), shouldThrow);
	}
	update(id, input, opts) {
		return this.#ctx.run(opts, (client) => updateProject({
			client,
			path: { project_id: id },
			body: { project: input },
			throwOnError: false
		}), (data) => data.project);
	}
	delete(id, opts) {
		return this.#ctx.run(opts, (client) => deleteProject({
			client,
			path: { project_id: id },
			throwOnError: false
		}), (data) => data.project);
	}
	async transfer(input, opts) {
		const shouldThrow = opts?.throwOnError ?? this.#ctx.defaults.throwOnError;
		const fromOrgId = input.fromOrgId ?? this.#ctx.defaults.orgId;
		if (!fromOrgId) return finalize(err(new NeonError("Pass fromOrgId or set orgId on the client.", "client")), shouldThrow);
		return this.#ctx.run(opts, (client) => transferProjectsFromOrgToOrg({
			client,
			path: { source_org_id: fromOrgId },
			body: {
				destination_org_id: input.toOrgId,
				project_ids: input.projectIds
			},
			throwOnError: false
		}), () => void 0);
	}
	transferFromUser(input, opts) {
		return this.#ctx.run(opts, (client) => transferProjectsFromUserToOrg({
			client,
			body: {
				destination_org_id: input.toOrgId,
				project_ids: input.projectIds
			},
			throwOnError: false
		}), () => void 0);
	}
};
//#endregion
export { Projects };
