// @neon/sdk/dist/neon/resources/roles.js
import { createProjectBranchRole, deleteProjectBranchRole, getProjectBranchRole, getProjectBranchRolePassword, listProjectBranchRoles, resetProjectBranchRolePassword } from "../../client/sdk.gen.js";
//#region src/neon/resources/roles.ts
/** Role resource (branch-scoped). */
var Roles = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	list(projectId, branchId, opts) {
		return this.#ctx.run(opts, (client) => listProjectBranchRoles({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			throwOnError: false
		}), (data) => data.roles);
	}
	get(projectId, branchId, name, opts) {
		return this.#ctx.run(opts, (client) => getProjectBranchRole({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				role_name: name
			},
			throwOnError: false
		}), (data) => data.role);
	}
	create(projectId, branchId, input, opts) {
		return this.#ctx.run(opts, (client) => createProjectBranchRole({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			body: { role: input },
			throwOnError: false
		}), (data) => data.role);
	}
	delete(projectId, branchId, name, opts) {
		return this.#ctx.runVoid(opts, (client) => deleteProjectBranchRole({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				role_name: name
			},
			throwOnError: false
		}));
	}
	password(projectId, branchId, name, opts) {
		return this.#ctx.run(opts, (client) => getProjectBranchRolePassword({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				role_name: name
			},
			throwOnError: false
		}), (data) => data.password);
	}
	resetPassword(projectId, branchId, name, opts) {
		return this.#ctx.run(opts, (client) => resetProjectBranchRolePassword({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId,
				role_name: name
			},
			throwOnError: false
		}), (data) => data.role);
	}
};
//#endregion
export { Roles };
