// @neon/sdk/dist/neon/resources/snapshots.js
import { err, finalize, ok } from "../result.js";
import { createSnapshot, deleteProjectBranch, deleteSnapshot, finalizeRestoreBranch, getSnapshotSchedule, listSnapshots, restoreSnapshot, setSnapshotSchedule, updateSnapshot } from "../../client/sdk.gen.js";
//#region src/neon/resources/snapshots.ts
/** Snapshot resource. */
var Snapshots = class {
	#ctx;
	constructor(ctx) {
		this.#ctx = ctx;
	}
	list(projectId, opts) {
		return this.#ctx.run(opts, (client) => listSnapshots({
			client,
			path: { project_id: projectId },
			throwOnError: false
		}), (data) => data.snapshots);
	}
	create(projectId, branchId, input, opts) {
		return this.#ctx.run(opts, (client) => createSnapshot({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			query: {
				name: input?.name,
				timestamp: input?.timestamp,
				lsn: input?.lsn,
				expires_at: input?.expiresAt
			},
			throwOnError: false
		}), (data) => data.snapshot);
	}
	update(projectId, snapshotId, input, opts) {
		return this.#ctx.run(opts, (client) => updateSnapshot({
			client,
			path: {
				project_id: projectId,
				snapshot_id: snapshotId
			},
			body: { snapshot: input },
			throwOnError: false
		}), (data) => data.snapshot);
	}
	delete(projectId, snapshotId, opts) {
		return this.#ctx.run(opts, (client) => deleteSnapshot({
			client,
			path: {
				project_id: projectId,
				snapshot_id: snapshotId
			},
			throwOnError: false
		}), () => void 0);
	}
	async restore(projectId, snapshotId, input, opts) {
		const shouldThrow = opts?.throwOnError ?? this.#ctx.defaults.throwOnError;
		const preview = input?.preview;
		const finalizeNow = preview ? false : input?.finalize ?? input?.targetBranchId === void 0;
		const restored = await this.#ctx.execute({
			...opts,
			waitForReadiness: preview ? true : opts?.waitForReadiness
		}, (client) => restoreSnapshot({
			client,
			path: {
				project_id: projectId,
				snapshot_id: snapshotId
			},
			body: {
				name: input?.name,
				target_branch_id: input?.targetBranchId,
				finalize_restore: finalizeNow
			},
			throwOnError: false
		}), (data) => data.branch);
		if (restored.error || !preview) return finalize(restored, shouldThrow);
		const branch = restored.data;
		const step = await preview(branch) ? await this.#ctx.execute({
			...opts,
			waitForReadiness: true
		}, (client) => finalizeRestoreBranch({
			client,
			path: {
				project_id: projectId,
				branch_id: branch.id
			},
			throwOnError: false
		}), () => void 0) : input?.keepOnAbort ? ok(void 0) : await this.#ctx.execute({
			...opts,
			waitForReadiness: true
		}, (client) => deleteProjectBranch({
			client,
			path: {
				project_id: projectId,
				branch_id: branch.id
			},
			throwOnError: false
		}), () => void 0);
		if (step.error) return finalize(err(step.error), shouldThrow);
		return finalize(ok(branch), shouldThrow);
	}
	getSchedule(projectId, branchId, opts) {
		return this.#ctx.run(opts, (client) => getSnapshotSchedule({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			throwOnError: false
		}), (data) => data);
	}
	setSchedule(projectId, branchId, schedule, opts) {
		return this.#ctx.run(opts, (client) => setSnapshotSchedule({
			client,
			path: {
				project_id: projectId,
				branch_id: branchId
			},
			body: schedule,
			throwOnError: false
		}), () => void 0);
	}
};
//#endregion
export { Snapshots };
