// @neon/config/dist/lib/define-config.js
import { ConfigValidationError } from "./errors.js";
import { parseBranchTtl } from "./duration.js";
import { branchTuningSchema, configInputSchema, formatZodIssues } from "./schema.js";
//#region src/lib/define-config.ts
/** Default deploy parameters applied to functions that omit them in `neon.ts`. */
const DEFAULT_FUNCTION_RUNTIME = "nodejs24";
const REGION_PREFIX = /^(aws|azure|gcp)-/;
/**
* Validate and freeze a Neon Platform branch policy.
*
* Used at the top of `neon.ts`:
* ```ts
* import { defineConfig } from "@neon/config/v1";
*
* export default defineConfig({
*   auth: true,
*   preview: {
*     functions: {
*       hello: { name: "Hello", source: "./functions/hello.ts", dev: { port: 8787 } },
*     },
*   },
*   branch: (branch) => ({ protected: branch.name === "main" }),
* });
* ```
*
* The policy is split into a **static** existential set (top-level `auth` / `dataApi`
* toggles and the beta `preview` block) and a **dynamic** per-branch `branch` closure. The
* static half determines which secrets exist — so `NeonEnv<typeof config>` and `parseEnv`
* are exact — while the closure can only *tune* a branch (lifecycle, compute, per-function
* deploy settings), never change what exists.
*
* The `branch` callback receives a read-only {@link BranchTarget} descriptor of the branch
* being decided for (not a live handle); switch on its facts (`branch.name`,
* `branch.isDefault`, `branch.exists`, …) and **return** the desired tuning. It runs in two
* modes: against an existing branch (fields populated from Neon) and during pre-create
* evaluation (`exists: false`, `id` undefined).
*
* Pure: no I/O, no side effects. The static parts are validated here; the closure's output
* is validated every time it is evaluated so errors point at the concrete branch target.
*/
function defineConfig(input) {
	if (typeof input === "function") throw new ConfigValidationError(["defineConfig now expects an object, not a function: `export default defineConfig({ auth: true, preview: { … }, branch: (branch) => ({ … }) })`.", "The static services/preview set moved to the top level; per-branch logic moved into the `branch` closure."]);
	if (input === null || typeof input !== "object") throw new ConfigValidationError(["defineConfig expects a configuration object: `export default defineConfig({ … })`."]);
	const parsed = configInputSchema.safeParse(input);
	if (!parsed.success) throw new ConfigValidationError(formatZodIssues(parsed.error));
	return Object.freeze({ ...input });
}
/**
* Evaluate a branch policy for a specific branch target and return a normalized config.
*
* Merges the static existential set (services + preview functions/buckets) with the
* per-branch tuning returned by the `branch` closure into the same {@link
* ResolvedBranchConfig} the rest of the runtime (diff / push / fetchEnv) consumes.
*/
function resolveConfig(config, branch) {
	const tuning = evaluateBranchTuning(config.branch, branch);
	const resolved = {
		authEnabled: isServiceEnabled(config.auth),
		dataApiEnabled: isDataApiEnabled(config.dataApi)
	};
	const dataApi = resolveDataApi(config.dataApi);
	if (dataApi) resolved.dataApi = dataApi;
	if (tuning.parent !== void 0) resolved.parent = tuning.parent;
	if (tuning.ttl !== void 0) {
		const parsedTtl = parseBranchTtl(tuning.ttl);
		if (!("error" in parsedTtl)) resolved.ttlSeconds = parsedTtl.seconds;
	}
	if (tuning.protected !== void 0) resolved.protected = tuning.protected;
	if (tuning.postgres?.computeSettings) resolved.postgres = { computeSettings: { ...tuning.postgres.computeSettings } };
	const preview = resolvePreviewConfig(config.preview, tuning);
	if (preview) resolved.preview = preview;
	return resolved;
}
/**
* Run the `branch` closure (when present) for the target and validate its output. The
* closure is optional — a fully static policy resolves with empty tuning.
*/
function evaluateBranchTuning(branchFn, target) {
	if (!branchFn) return {};
	let raw;
	try {
		raw = branchFn(Object.freeze({ ...target }));
	} catch (cause) {
		throw new ConfigValidationError([`Branch policy threw while evaluating branch "${target.name}".`, cause?.message ?? String(cause)]);
	}
	const parsed = branchTuningSchema.safeParse(raw ?? {});
	if (!parsed.success) throw new ConfigValidationError(formatZodIssues(parsed.error));
	return parsed.data;
}
function isServiceEnabled(toggle) {
	if (toggle === void 0) return false;
	if (typeof toggle === "boolean") return toggle;
	return toggle.enabled !== false;
}
/** Whether a {@link DataApiInput} is enabled (present object/`true` unless `enabled: false`). */
function isDataApiEnabled(input) {
	if (input === void 0) return false;
	if (typeof input === "boolean") return input;
	return input.enabled !== false;
}
/**
* Normalize a {@link DataApiInput} into a {@link ResolvedDataApiConfig}, or `undefined` when
* the Data API is not enabled. `authProvider` defaults to `"neon"`; the external-IdP wiring
* is carried through only for the `"external"` provider; `settings` is copied with its
* `undefined` entries dropped so diffing only considers fields the policy actually set.
*/
function resolveDataApi(input) {
	if (!isDataApiEnabled(input)) return void 0;
	if (typeof input !== "object") return { authProvider: "neon" };
	const authProvider = input.authProvider ?? "neon";
	const resolved = { authProvider };
	if (authProvider === "external") {
		if (input.jwksUrl !== void 0) resolved.jwksUrl = input.jwksUrl;
		if (input.providerName !== void 0) resolved.providerName = input.providerName;
		if (input.jwtAudience !== void 0) resolved.jwtAudience = input.jwtAudience;
	}
	const settings = normalizeDataApiSettings(input.settings);
	if (settings) resolved.settings = settings;
	return resolved;
}
/** Copy a {@link DataApiSettings}, dropping `undefined` entries; `undefined` when empty. */
function normalizeDataApiSettings(settings) {
	if (!settings) return void 0;
	const out = {};
	for (const [key, value] of Object.entries(settings)) if (value !== void 0) out[key] = value;
	return Object.keys(out).length > 0 ? out : void 0;
}
/**
* Normalize the static {@link PreviewInput} (merged with per-branch function tuning) into a
* {@link ResolvedPreviewConfig}. Returns `undefined` when the policy declares no `preview`
* block so the field can be omitted entirely. Function slugs / bucket names come from the
* record keys.
*/
function resolvePreviewConfig(preview, tuning) {
	if (!preview) return void 0;
	const fnTuning = tuning.preview?.functions ?? {};
	return {
		functions: Object.entries(preview.functions ?? {}).map(([slug, def]) => resolveFunctionConfig(slug, def, fnTuning[slug] ?? {})),
		buckets: Object.entries(preview.buckets ?? {}).map(([name, def]) => ({
			name,
			access: def.access ?? "private"
		})),
		aiGatewayEnabled: isServiceEnabled(preview.aiGateway)
	};
}
function resolveFunctionConfig(slug, def, tuning) {
	return {
		slug,
		name: def.name,
		source: def.source,
		env: { ...def.env ?? {} },
		runtime: tuning.runtime ?? DEFAULT_FUNCTION_RUNTIME,
		...def.dev ? { dev: def.dev } : {}
	};
}
/**
* Normalize a region identifier to Neon's `<cloud>-<region>` format. When the user writes
* `us-east-1` we assume `aws-us-east-1`. Pure helper used by both the validator and the
* NeonApi adapter.
*/
function normalizeRegion(region) {
	if (REGION_PREFIX.test(region)) return region;
	return `aws-${region}`;
}
//#endregion
export { defineConfig, normalizeRegion, resolveConfig };
