// @neon/config/dist/lib/duration.js
//#region src/lib/duration.ts
/**
* Parse a duration value into whole seconds.
*
* Accepted formats:
* - a positive finite **number** → interpreted as seconds (must be an integer)
* - a **string** of the form `<integer><unit>` where unit is one of `s`, `m`, `h`, `d`, `w`
*   (e.g. `30s`, `5m`, `1h`, `7d`, `2w`)
*
* A **unit is required** on strings: a bare numeric string like `"7"` is rejected — pass a
* `number` (`7`) for raw seconds instead. This removes the ambiguity where `"7"` silently
* meant 7 seconds rather than, say, `"7d"`.
*
* Returns `{ seconds }` on success or `{ error }` on failure. Pure function — never throws.
*/
function parseDuration(input) {
	if (typeof input === "number") {
		if (!Number.isFinite(input)) return { error: `not a finite number: ${input}` };
		if (!Number.isInteger(input)) return { error: `must be an integer when passed as number: ${input}` };
		if (input <= 0) return { error: `must be > 0, got ${input}` };
		return { seconds: input };
	}
	const trimmed = input.trim();
	if (trimmed === "") return { error: "duration string is empty" };
	if (/^\d+$/.test(trimmed)) return { error: `duration string "${input}" is missing a unit; add one of s, m, h, d, w (e.g. "${trimmed}d") or pass ${trimmed} as a number for seconds` };
	const unitMatch = /^(\d+)([smhdw])$/i.exec(trimmed);
	if (!unitMatch) return { error: `invalid duration "${input}"; expected an integer followed by one of: s, m, h, d, w (e.g. "30s", "1h", "7d")` };
	const value = Number(unitMatch[1]);
	const unit = unitMatch[2].toLowerCase();
	if (value <= 0) return { error: `must be > 0, got "${trimmed}"` };
	return { seconds: value * UNIT_SECONDS[unit] };
}
const UNIT_SECONDS = {
	s: 1,
	m: 60,
	h: 3600,
	d: 1440 * 60,
	w: 10080 * 60
};
/** Neon's branch-expiration ceiling: the API rejects an `expires_at` more than 30 days out. */
const MAX_BRANCH_TTL_SECONDS = 30 * UNIT_SECONDS.d;
/**
* Parse a branch TTL into seconds, enforcing Neon's branch-expiration limit on top of the
* shared {@link parseDuration} rules: the result must be `> 0` and at most 30 days
* ({@link MAX_BRANCH_TTL_SECONDS}), since the API caps `expires_at` at 30 days from now.
*
* Returns `{ seconds }` on success or `{ error }` on failure. Pure function — never throws.
*/
function parseBranchTtl(input) {
	const result = parseDuration(input);
	if ("error" in result) return result;
	if (result.seconds > MAX_BRANCH_TTL_SECONDS) return { error: `branch TTL must be at most 30 days (${MAX_BRANCH_TTL_SECONDS}s), got ${result.seconds}s` };
	return result;
}
/**
* Render a TTL in seconds back to the canonical "<n><unit>" form. Used for round-trip
* serialization when {@link pullConfig} emits a TTL value (it always falls back to seconds
* when no clean unit boundary matches). The output always carries a unit, so it is a valid
* {@link DurationString}.
*/
function formatDurationSeconds(totalSeconds) {
	if (!Number.isFinite(totalSeconds) || totalSeconds <= 0) throw new RangeError(`formatDurationSeconds expected a positive finite number, got ${totalSeconds}`);
	const candidates = [
		["w", UNIT_SECONDS.w],
		["d", UNIT_SECONDS.d],
		["h", UNIT_SECONDS.h],
		["m", UNIT_SECONDS.m]
	];
	for (const [unit, perUnit] of candidates) if (totalSeconds % perUnit === 0) return `${totalSeconds / perUnit}${unit}`;
	return `${totalSeconds}s`;
}
/**
* Parse a suspend timeout value into seconds for the Neon API.
*
* Accepted formats:
* - `false` → -1 (never suspend)
* - `undefined` → 0 (use platform default)
* - duration string → parsed seconds ("5m", "1h", "7d")
* - number → validated seconds (must be 60-604800 or -1/0)
*
* Returns `{ seconds }` on success or `{ error }` on failure. Pure function — never throws.
*/
function parseSuspendTimeout(input) {
	if (input === false) return { seconds: -1 };
	if (input === void 0) return { seconds: 0 };
	if (typeof input === "number") {
		if (!Number.isFinite(input)) return { error: `not a finite number: ${input}` };
		if (!Number.isInteger(input)) return { error: `must be an integer: ${input}` };
		if (input === -1 || input === 0) return { seconds: input };
		if (input < 60 || input > 604800) return { error: `suspend timeout must be between 60 and 604800 seconds (1 minute to 1 week), got ${input}` };
		return { seconds: input };
	}
	const result = parseDuration(input);
	if ("error" in result) return result;
	const { seconds } = result;
	if (seconds < 60 || seconds > 604800) return { error: `suspend timeout must be between 60 and 604800 seconds (1 minute to 1 week), "${input}" = ${seconds}s` };
	return { seconds };
}
/**
* Format a suspend timeout value from API seconds back to the user-facing type.
* Returns `false` for -1 (never suspend), `undefined` for 0 (default), or a duration string.
*/
function formatSuspendTimeout(seconds) {
	if (seconds === -1) return false;
	if (seconds === 0) return void 0;
	return formatDurationSeconds(seconds);
}
//#endregion
export { MAX_BRANCH_TTL_SECONDS, formatDurationSeconds, formatSuspendTimeout, parseBranchTtl, parseDuration, parseSuspendTimeout };
