// @neon/config/dist/lib/loader.js
import { ConfigLoadError, isPlatformError } from "./errors.js";
import { defineConfig } from "./define-config.js";
import { existsSync, statSync } from "node:fs";
import { dirname, isAbsolute, resolve } from "node:path";
import { homedir } from "node:os";
import { pathToFileURL } from "node:url";
//#region src/lib/loader.ts
/**
* Default file names tried (in order) when {@link loadConfigFromFile} is called without an
* explicit path. We accept `.ts` first because that's the documented format; `.mjs` and `.js`
* fall out for free since jiti handles all of them.
*/
const DEFAULT_CONFIG_FILENAMES = [
	"neon.ts",
	"neon.mts",
	"neon.js",
	"neon.mjs"
];
/**
* Load a `neon.ts` (or any other supported extension) and return the validated {@link Config}.
*
* Behavior:
* - When `path` is set, that file is loaded directly. The file must exist and must default-export
*   a value produced by `defineConfig()`.
* - When `path` is omitted, we walk up from `cwd` picking the **closest** file matching
*   {@link DEFAULT_CONFIG_FILENAMES}. The walk is monorepo-friendly: intermediate
*   `package.json` files do **not** stop it, so a single `neon.ts` lifted to the workspace
*   root keeps working when invoked from inside any sub-package. The walk terminates at the
*   first directory containing `.git`, at `stopAt`, or at the filesystem root.
*
* jiti is loaded lazily so that callers who pass an already-resolved `Config` to `pushConfig`
* never pay the import cost.
*/
async function loadConfigFromFile(options = {}) {
	const resolvedPath = options.path ? resolveExplicitPath(options.path, options.cwd) : findDefaultConfig(options.cwd, options.stopAt);
	if (!resolvedPath) throw new ConfigLoadError([
		`Could not find a Neon config file while walking up from ${resolve(options.cwd ?? process.cwd())}.`,
		`Looked for: ${DEFAULT_CONFIG_FILENAMES.join(", ")} (stopping at the first directory with a \`.git\`).`,
		`Create one at your repository root (or anywhere on the path from cwd up to .git), or pass an explicit \`configPath\` (SDK) / \`--config <path>\` (CLI).`
	].join("\n"));
	let mod;
	try {
		mod = await importModule(resolvedPath);
	} catch (cause) {
		if (isPlatformError(cause)) throw cause;
		throw new ConfigLoadError([
			`Failed to evaluate ${resolvedPath}.`,
			`Underlying error: ${cause instanceof Error ? cause.message : String(cause)}`,
			"This is usually a TypeScript syntax error, a missing dependency, or a runtime exception inside the config file. Run the file directly (e.g. `npx tsx neon.ts`) to reproduce."
		].join("\n"), { cause });
	}
	const exported = extractDefaultExport(mod);
	if (exported === void 0) throw new ConfigLoadError([`${resolvedPath} loaded successfully but did not default-export a config.`, "Add `export default defineConfig({ ... })` at the bottom of the file. (Named exports are ignored.)"].join("\n"));
	return {
		config: defineConfig(exported),
		resolvedPath
	};
}
function resolveExplicitPath(input, cwd) {
	const base = resolve(cwd ?? process.cwd());
	const abs = isAbsolute(input) ? input : resolve(base, input);
	if (!existsSync(abs)) throw new ConfigLoadError(`Config file not found at ${abs}. The path was resolved from \`${input}\` against ${base}.`);
	if (!statSync(abs).isFile()) throw new ConfigLoadError(`Config path ${abs} is a directory, not a file. Pass a path to the file itself (e.g. ./neon.ts).`);
	return abs;
}
function findDefaultConfig(cwd, stopAt) {
	let current = resolve(cwd ?? process.cwd());
	const stop = resolve(stopAt ?? homedir());
	let lastSeen = null;
	while (true) {
		for (const name of DEFAULT_CONFIG_FILENAMES) {
			const candidate = resolve(current, name);
			if (existsSync(candidate) && safeIsFile(candidate)) return candidate;
		}
		if (existsSync(resolve(current, ".git"))) return null;
		if (current === stop) return null;
		const parent = dirname(current);
		if (parent === current || parent === lastSeen) return null;
		lastSeen = current;
		current = parent;
	}
}
async function importModule(absPath) {
	const lower = absPath.toLowerCase();
	if (!(lower.endsWith(".ts") || lower.endsWith(".mts") || lower.endsWith(".cts"))) return import(pathToFileURL(absPath).href);
	const createJiti = extractCreateJiti(await import("jiti"));
	if (!createJiti) throw new ConfigLoadError(["jiti is required to load TypeScript config files but could not be initialised.", "Reinstall the package dependencies (`pnpm install` / `npm install`) — jiti is a runtime dependency of @neon/config."].join(" "));
	return createJiti(pathToFileURL(absPath).href, {
		interopDefault: true,
		moduleCache: false
	}).import(absPath);
}
function extractCreateJiti(mod) {
	if (mod === null || typeof mod !== "object") return null;
	const obj = mod;
	if (typeof obj.createJiti === "function") return obj.createJiti;
	const def = obj.default;
	if (def !== null && typeof def === "object") {
		const defObj = def;
		if (typeof defObj.createJiti === "function") return defObj.createJiti;
	}
	return null;
}
function extractDefaultExport(mod) {
	if (mod === null || typeof mod !== "object") return mod;
	const obj = mod;
	if ("default" in obj && obj.default !== void 0) return obj.default;
	if (typeof mod === "function") return mod;
}
function safeIsFile(path) {
	try {
		return statSync(path).isFile();
	} catch {
		return false;
	}
}
//#endregion
export { DEFAULT_CONFIG_FILENAMES, loadConfigFromFile };
