// @neon/sdk/dist/client/client/client.gen.js
import { createSseClient } from "../core/serverSentEvents.gen.js";
import { getValidRequestBody } from "../core/utils.gen.js";
import { buildUrl, createConfig, createInterceptors, getParseAs, mergeConfigs, mergeHeaders, setAuthParams } from "./utils.gen.js";
//#region src/client/client/client.gen.ts
const createClient = (config = {}) => {
	let _config = mergeConfigs(createConfig(), config);
	const getConfig = () => ({ ..._config });
	const setConfig = (config) => {
		_config = mergeConfigs(_config, config);
		return getConfig();
	};
	const interceptors = createInterceptors();
	const beforeRequest = async (options) => {
		const opts = {
			..._config,
			...options,
			fetch: options.fetch ?? _config.fetch ?? globalThis.fetch,
			headers: mergeHeaders(_config.headers, options.headers),
			serializedBody: void 0
		};
		if (opts.security) await setAuthParams(opts);
		if (opts.requestValidator) await opts.requestValidator(opts);
		if (opts.body !== void 0 && opts.bodySerializer) opts.serializedBody = opts.bodySerializer(opts.body);
		if (opts.body === void 0 || opts.serializedBody === "") opts.headers.delete("Content-Type");
		const resolvedOpts = opts;
		return {
			opts: resolvedOpts,
			url: buildUrl(resolvedOpts)
		};
	};
	const request = async (options) => {
		const throwOnError = options.throwOnError ?? _config.throwOnError;
		const responseStyle = options.responseStyle ?? _config.responseStyle;
		let request;
		let response;
		try {
			const { opts, url } = await beforeRequest(options);
			const requestInit = {
				redirect: "follow",
				...opts,
				body: getValidRequestBody(opts)
			};
			request = new Request(url, requestInit);
			for (const fn of interceptors.request.fns) if (fn) request = await fn(request, opts);
			const _fetch = opts.fetch;
			response = await _fetch(request);
			for (const fn of interceptors.response.fns) if (fn) response = await fn(response, request, opts);
			const result = {
				request,
				response
			};
			if (response.ok) {
				const parseAs = (opts.parseAs === "auto" ? getParseAs(response.headers.get("Content-Type")) : opts.parseAs) ?? "json";
				if (response.status === 204 || response.headers.get("Content-Length") === "0") {
					let emptyData;
					switch (parseAs) {
						case "arrayBuffer":
						case "blob":
						case "text":
							emptyData = await response[parseAs]();
							break;
						case "formData":
							emptyData = new FormData();
							break;
						case "stream":
							emptyData = response.body;
							break;
						default:
							emptyData = {};
							break;
					}
					return opts.responseStyle === "data" ? emptyData : {
						data: emptyData,
						...result
					};
				}
				let data;
				switch (parseAs) {
					case "arrayBuffer":
					case "blob":
					case "formData":
					case "text":
						data = await response[parseAs]();
						break;
					case "json": {
						const text = await response.text();
						data = text ? JSON.parse(text) : {};
						break;
					}
					case "stream": return opts.responseStyle === "data" ? response.body : {
						data: response.body,
						...result
					};
				}
				if (parseAs === "json") {
					if (opts.responseValidator) await opts.responseValidator(data);
					if (opts.responseTransformer) data = await opts.responseTransformer(data);
				}
				return opts.responseStyle === "data" ? data : {
					data,
					...result
				};
			}
			const textError = await response.text();
			let jsonError;
			try {
				jsonError = JSON.parse(textError);
			} catch {}
			throw jsonError ?? textError;
		} catch (error) {
			let finalError = error;
			for (const fn of interceptors.error.fns) if (fn) finalError = await fn(finalError, response, request, options);
			finalError = finalError || {};
			if (throwOnError) throw finalError;
			return responseStyle === "data" ? void 0 : {
				error: finalError,
				request,
				response
			};
		}
	};
	const makeMethodFn = (method) => (options) => request({
		...options,
		method
	});
	const makeSseFn = (method) => async (options) => {
		const { opts, url } = await beforeRequest(options);
		return createSseClient({
			...opts,
			body: opts.body,
			method,
			onRequest: async (url, init) => {
				let request = new Request(url, init);
				for (const fn of interceptors.request.fns) if (fn) request = await fn(request, opts);
				return request;
			},
			serializedBody: getValidRequestBody(opts),
			url
		});
	};
	const _buildUrl = (options) => buildUrl({
		..._config,
		...options
	});
	return {
		buildUrl: _buildUrl,
		connect: makeMethodFn("CONNECT"),
		delete: makeMethodFn("DELETE"),
		get: makeMethodFn("GET"),
		getConfig,
		head: makeMethodFn("HEAD"),
		interceptors,
		options: makeMethodFn("OPTIONS"),
		patch: makeMethodFn("PATCH"),
		post: makeMethodFn("POST"),
		put: makeMethodFn("PUT"),
		request,
		setConfig,
		sse: {
			connect: makeSseFn("CONNECT"),
			delete: makeSseFn("DELETE"),
			get: makeSseFn("GET"),
			head: makeSseFn("HEAD"),
			options: makeSseFn("OPTIONS"),
			patch: makeSseFn("PATCH"),
			post: makeSseFn("POST"),
			put: makeSseFn("PUT"),
			trace: makeSseFn("TRACE")
		},
		trace: makeMethodFn("TRACE")
	};
};
//#endregion
export { createClient };
