// @neon/sdk/dist/neon/client.js
import { resolveConfig } from "./config.js";
import { RequestContext } from "./context.js";
import { ApiKeys, Regions, User } from "./resources/account.js";
import { Branches } from "./resources/branches.js";
import { Consumption } from "./resources/consumption.js";
import { Operations } from "./resources/operations.js";
import { Postgres } from "./resources/postgres.js";
import { Projects } from "./resources/projects.js";
import { Snapshots } from "./resources/snapshots.js";
//#region src/neon/client.ts
/**
* Create an ergonomic Neon client.
*
* @example
* ```ts
* import { createNeonClient } from "@neon/sdk";
*
* const neon = createNeonClient({ apiKey: process.env.NEON_API_KEY! });
* const { data, error } = await neon.projects.list().all();
* ```
*/
function createNeonClient(config) {
	const ctx = new RequestContext(resolveConfig(config));
	return {
		projects: new Projects(ctx),
		branches: new Branches(ctx),
		postgres: new Postgres(ctx),
		snapshots: new Snapshots(ctx),
		operations: new Operations(ctx),
		consumption: new Consumption(ctx),
		apiKeys: new ApiKeys(ctx),
		regions: new Regions(ctx),
		user: new User(ctx),
		client: ctx.client
	};
}
//#endregion
export { createNeonClient };
