// @neon/sdk/dist/neon/context.js
import { toNeonError } from "./errors.js";
import { err, finalize, ok } from "./result.js";
import { withRetries } from "./retry.js";
import { hasOperations, waitForOperations } from "./wait.js";
//#region src/neon/context.ts
/**
* Shared execution core: runs a raw client call with retries, maps its envelope to the
* ergonomic resource shape, optionally waits for provisioning operations, and applies the
* resolved `throwOnError` policy. Returns the bare value (throwing) or a {@link NeonResult}.
*/
var RequestContext = class {
	#config;
	constructor(config) {
		this.#config = config;
	}
	get client() {
		return this.#config.client;
	}
	get defaults() {
		return this.#config;
	}
	/** Run a raw call and map its body; applies the resolved `throwOnError` policy. */
	async run(opts, exec, map) {
		const shouldThrow = opts?.throwOnError ?? this.#config.throwOnError;
		return finalize(await this.execute(opts, exec, map), shouldThrow);
	}
	/** Like {@link run} but for endpoints that may return an empty (204) body. */
	async runVoid(opts, exec) {
		const shouldThrow = opts?.throwOnError ?? this.#config.throwOnError;
		const raw = await withRetries(() => exec(this.#config.client), this.#config.retries, opts?.signal);
		if (raw.error !== void 0) return finalize(err(toNeonError(raw.error, raw.response)), shouldThrow);
		const readiness = await this.#maybeWait(opts, raw.data);
		if (readiness?.error) return finalize(err(readiness.error), shouldThrow);
		return finalize(ok(void 0), shouldThrow);
	}
	/**
	* Run a raw call and map its body, returning the {@link NeonResult} envelope (no
	* throw). Used by `run` and by workflows that post-process the result.
	*/
	async execute(opts, exec, map) {
		const raw = await withRetries(() => exec(this.#config.client), this.#config.retries, opts?.signal);
		if (raw.error || raw.data === void 0) return err(toNeonError(raw.error, raw.response));
		const readiness = await this.#maybeWait(opts, raw.data);
		if (readiness?.error) return err(readiness.error);
		return ok(map(raw.data));
	}
	/** Poll provisioning operations when `waitForReadiness` is on and the body has any. */
	async #maybeWait(opts, data) {
		if (!(opts?.waitForReadiness ?? this.#config.waitForReadiness) || !hasOperations(data)) return void 0;
		return waitForOperations(this.#config.client, data.operations, {
			...this.#config.waitOptions,
			signal: opts?.signal ?? this.#config.waitOptions.signal
		});
	}
};
//#endregion
export { RequestContext };
