// @neon/sdk/dist/neon/errors.js
//#region src/neon/errors.ts
/** Base class for every error the ergonomic layer produces. */
var NeonError = class extends Error {
	kind;
	constructor(message, kind, options) {
		super(message, options);
		this.name = new.target.name;
		this.kind = kind;
	}
};
/** A non-2xx HTTP response from the Neon API. */
var NeonApiError = class extends NeonError {
	/** HTTP status code. */
	status;
	/** Machine-readable Neon error code (`GeneralError.code`), when present. */
	code;
	/** Neon request id (`X-Request-Id` / `GeneralError.request_id`), when present. */
	requestId;
	/** The raw response, when one was received. */
	response;
	/** The parsed error body, as returned by the API. */
	body;
	constructor(message, init) {
		super(message, init.kind ?? "api");
		this.status = init.status;
		this.code = init.code;
		this.requestId = init.requestId;
		this.response = init.response;
		this.body = init.body;
	}
};
/** 404 — the resource does not exist. */
var NeonNotFoundError = class extends NeonApiError {
	constructor(message, init) {
		super(message, {
			...init,
			kind: "not_found"
		});
	}
};
/** 401/403 — the API key is missing, invalid, or lacks permission. */
var NeonAuthError = class extends NeonApiError {
	constructor(message, init) {
		super(message, {
			...init,
			kind: "auth"
		});
	}
};
/** 429 — rate limited (after retries, if enabled, were exhausted). */
var NeonRateLimitError = class extends NeonApiError {
	constructor(message, init) {
		super(message, {
			...init,
			kind: "rate_limit"
		});
	}
};
/** An awaited Neon operation ended in a non-success terminal state. */
var NeonOperationError = class extends NeonError {
	/** The id of the operation that failed. */
	operationId;
	/** The terminal status reported by the API (`failed` / `error` / `cancelled`). */
	status;
	constructor(message, init) {
		super(message, "operation");
		this.operationId = init.operationId;
		this.status = init.status;
	}
};
/** Waiting for operations to finish exceeded the configured timeout. */
var NeonTimeoutError = class extends NeonError {
	constructor(message) {
		super(message, "timeout");
	}
};
/** A transport-level failure (DNS, connection, abort) — no HTTP response received. */
var NeonNetworkError = class extends NeonError {
	constructor(message, options) {
		super(message, "network", options);
	}
};
function readApiErrorBody(body) {
	if (typeof body !== "object" || body === null) return {};
	const out = {};
	if ("message" in body && typeof body.message === "string") out.message = body.message;
	if ("code" in body && typeof body.code === "string") out.code = body.code;
	if ("request_id" in body && typeof body.request_id === "string") out.request_id = body.request_id;
	return out;
}
/**
* Build the right {@link NeonError} subclass from a raw client result. `error` is the
* decoded error body (Neon `GeneralError`); `response` is present unless the failure was
* transport-level.
*/
function toNeonError(error, response) {
	if (!response) return new NeonNetworkError("Network error: no response received from the Neon API.", { cause: error });
	const parsed = readApiErrorBody(error);
	const status = response.status;
	const message = parsed.message ?? `Neon API request failed with status ${status}.`;
	const init = {
		status,
		code: parsed.code,
		requestId: parsed.request_id ?? response.headers.get("x-request-id") ?? void 0,
		response,
		body: error
	};
	if (status === 404) return new NeonNotFoundError(message, init);
	if (status === 401 || status === 403) return new NeonAuthError(message, init);
	if (status === 429) return new NeonRateLimitError(message, init);
	return new NeonApiError(message, init);
}
//#endregion
export { NeonApiError, NeonAuthError, NeonError, NeonNetworkError, NeonNotFoundError, NeonOperationError, NeonRateLimitError, NeonTimeoutError, toNeonError };
