// @neon/sdk/dist/neon/wait.js
import { NeonOperationError, NeonTimeoutError, toNeonError } from "./errors.js";
import { err, ok } from "./result.js";
import { getProjectOperation } from "../client/sdk.gen.js";
//#region src/neon/wait.ts
const SUCCESS = /* @__PURE__ */ new Set(["finished", "skipped"]);
const FAILURE = /* @__PURE__ */ new Set([
	"failed",
	"error",
	"cancelled"
]);
const sleep = (ms, signal) => new Promise((resolve, reject) => {
	if (signal?.aborted) return reject(signal.reason);
	const id = setTimeout(resolve, ms);
	signal?.addEventListener("abort", () => {
		clearTimeout(id);
		reject(signal.reason);
	}, { once: true });
});
/**
* Poll the given operations until each reaches a terminal `finished`/`skipped` state.
* Returns an error result carrying {@link NeonOperationError} if any operation ends in
* `failed`/`error`/`cancelled`, or {@link NeonTimeoutError} if the deadline is exceeded.
*
* Composable: usable directly with operations obtained from the raw layer.
*/
async function waitForOperations(client, operations, options = {}) {
	const pollIntervalMs = options.pollIntervalMs ?? 1e3;
	const timeoutMs = options.timeoutMs ?? 3e5;
	const deadline = Date.now() + timeoutMs;
	let pending = operations.filter((op) => !SUCCESS.has(op.status));
	for (const op of pending) if (FAILURE.has(op.status)) return err(operationFailed(op));
	pending = pending.filter((op) => !FAILURE.has(op.status));
	while (pending.length > 0) {
		if (Date.now() > deadline) return err(new NeonTimeoutError(`Timed out after ${timeoutMs}ms waiting for ${pending.length} operation(s) to finish.`));
		await sleep(pollIntervalMs, options.signal);
		const stillPending = [];
		for (const op of pending) {
			const { data, error, response } = await getProjectOperation({
				client,
				path: {
					project_id: op.project_id,
					operation_id: op.id
				},
				throwOnError: false,
				signal: options.signal
			});
			if (error || !data) return err(toNeonError(error, response));
			const current = data.operation;
			if (FAILURE.has(current.status)) return err(operationFailed(current));
			if (!SUCCESS.has(current.status)) stillPending.push(current);
		}
		pending = stillPending;
	}
	return ok(void 0);
}
function operationFailed(op) {
	const detail = op.error ? `: ${op.error}` : "";
	return new NeonOperationError(`Operation ${op.id} (${op.action}) ended with status "${op.status}"${detail}.`, {
		operationId: op.id,
		status: op.status
	});
}
/** Type guard: does a response body carry an `operations` array we can wait on? */
function hasOperations(data) {
	return typeof data === "object" && data !== null && "operations" in data && Array.isArray(data.operations);
}
//#endregion
export { hasOperations, waitForOperations };
