// @neon/config/dist/v1.js
import { ConfigLoadError, ConfigValidationError, ErrorCode, MissingContextError, PlatformError, PushAbortedError, PushConflictError, isPlatformError } from "./lib/errors.js";
import { branchTuningSchema, bucketDefSchema, computeSettingsSchema, configInputSchema, dataApiConfigSchema, dataApiInputSchema, dataApiSettingsSchema, functionDefSchema, functionTuningSchema, postgresConfigSchema, previewInputSchema, serviceToggleInputSchema, serviceToggleSchema } from "./lib/schema.js";
import { createRealNeonApi } from "./lib/neon-api-real.js";
import { createNeonApiFromOptions, resolveApiKey } from "./lib/auth.js";
import { credentialScopesSatisfied, deriveCredentialScopes } from "./lib/credentials.js";
import { defineConfig, resolveConfig } from "./lib/define-config.js";
import { diffConfig } from "./lib/diff.js";
import { loadConfigFromFile } from "./lib/loader.js";
import { DATA_API_AUTH_PROVIDERS } from "./lib/types.js";
//#region src/v1.ts
/**
* `@neon/config/v1` — the v1 public API for Config-as-Code on the Neon Platform.
*
* Usage in `neon.ts`:
* ```ts
* import { defineConfig } from "@neon/config/v1";
*
* export default defineConfig({
*   // Static: what *exists* on every branch (drives the typed env).
*   auth: true,
*   // Dynamic: per-branch tuning only — cannot add/remove services.
*   branch: (branch) => ({
*     protected: branch.name === "main",
*     ...(branch.name === "main" ? {} : { parent: "main", ttl: "7d" }),
*   }),
* });
* ```
*
* This is the **authoring** surface — `defineConfig`, types, schemas, the pure diff engine,
* and the Neon API adapter. It is intentionally free of heavy/native dependencies so that
* importing it from `neon.ts` stays cheap and bundler-safe.
*
* The imperative operations (`inspect` / `plan` / `apply`, `pushConfig` / `pullConfig`) and
* function bundling/deploy live in **`@neon/config-runtime`**, which depends on this
* package and pulls in `esbuild`. Import that from your CLI / CI, not from `neon.ts`:
* ```ts
* import config from "../neon";
* import { inspect, plan, apply } from "@neon/config-runtime/v1";
* ```
*
* Surface guidelines:
* - Top-level: `defineConfig` / `resolveConfig`, the pure `diffConfig` engine, the
*   `createRealNeonApi` adapter + `NeonApi` types, the config loader, the `PlatformError`
*   base class + `ErrorCode` enum, and the config types used in `neon.ts`.
* - `errors` namespace: specific `PlatformError` subclasses (`ConfigLoadError`,
*   `PushConflictError`, …).
* - `schemas` namespace: the zod schemas underlying `defineConfig`.
*/
/**
* Specific `PlatformError` subclasses, grouped for `instanceof` / structured access.
* Also available as top-level exports.
*/
const errors = {
	ConfigLoadError,
	ConfigValidationError,
	ErrorCode,
	isPlatformError,
	MissingContextError,
	PlatformError,
	PushAbortedError,
	PushConflictError
};
/** The zod schemas underlying `defineConfig`, grouped under product-friendly names. */
const schemas = {
	config: configInputSchema,
	branchTuning: branchTuningSchema,
	bucket: bucketDefSchema,
	computeSettings: computeSettingsSchema,
	dataApi: dataApiConfigSchema,
	dataApiInput: dataApiInputSchema,
	dataApiSettings: dataApiSettingsSchema,
	function: functionDefSchema,
	functionTuning: functionTuningSchema,
	postgres: postgresConfigSchema,
	preview: previewInputSchema,
	service: serviceToggleSchema,
	serviceInput: serviceToggleInputSchema
};
//#endregion
export { ConfigLoadError, ConfigValidationError, DATA_API_AUTH_PROVIDERS, ErrorCode, MissingContextError, PlatformError, PushAbortedError, PushConflictError, createNeonApiFromOptions, createRealNeonApi, credentialScopesSatisfied, defineConfig, deriveCredentialScopes, diffConfig, errors, isPlatformError, loadConfigFromFile, resolveApiKey, resolveConfig, schemas };
