// @neon/sdk/dist/client/sdk.gen.js
import { client } from "./client.gen.js";
//#region src/client/sdk.gen.ts
/**
* Get advisor issues
*
* Analyzes the database for security and performance issues.
* Returns a list of issues categorized by severity (ERROR, WARN, INFO).
*
* Requires read access to the project and Data API enabled.
*
*/
const getProjectAdvisorSecurityIssues = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/advisors",
	...options
});
/**
* List API keys
*
* Retrieves the API keys for your Neon account.
* The response does not include API key tokens. A token is only provided when creating an API key.
* API keys can also be managed in the Neon Console.
* For more information, see [Manage API keys](https://neon.com/docs/manage/api-keys/).
*
*/
const listApiKeys = (options) => (options?.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/api_keys",
	...options
});
/**
* Create API key
*
* Creates an API key.
* The `key_name` is a user-specified name for the key.
* Returns an `id` and `key`; the `key` is a randomly generated, 64-bit token required to access the Neon API.
* Store the key securely — it is only returned once.
* API keys can also be managed in the Neon Console.
* See [Manage API keys](https://neon.com/docs/manage/api-keys/).
*
*/
const createApiKey = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/api_keys",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Revoke API key
*
* Revokes the specified API key.
* An API key that is no longer needed can be revoked.
* This action cannot be reversed.
* API keys can also be managed in the Neon Console.
* See [Manage API keys](https://neon.com/docs/manage/api-keys/).
*
*/
const revokeApiKey = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/api_keys/{key_id}",
	...options
});
/**
* Retrieve operation details
*
* Retrieves details for the specified operation.
* An operation is an action performed on a Neon project resource.
*
*/
const getProjectOperation = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/operations/{operation_id}",
	...options
});
/**
* List projects
*
* Retrieves a list of projects for the specified organization.
* If using a personal API key, include the `org_id` parameter to specify which organization to work with.
* If using an org API key, `org_id` is automatically inferred from the key.
* For more information, see [Manage organizations using the Neon API](https://neon.com/docs/manage/orgs-api)
* and [Manage projects](https://neon.com/docs/manage/projects/).
*
*/
const listProjects = (options) => (options?.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects",
	...options
});
/**
* Create project
*
* Creates a Neon project within an organization.
* If using a personal API key, include the `org_id` parameter to specify which organization to create the project in.
* If using an org API key, `org_id` is automatically inferred from the key.
* Plan limits define how many projects you can create.
* For more information, see [Manage projects](https://neon.com/docs/manage/projects/).
*
* You can specify a region and Postgres version in the request body.
* Neon currently supports PostgreSQL 14, 15, 16, 17, and 18.
* For supported regions and `region_id` values, see [Regions](https://neon.com/docs/introduction/regions/).
*
*/
const createProject = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List shared projects
*
* Retrieves a list of projects shared with your Neon account.
* For more information, see [Manage projects](https://neon.com/docs/manage/projects/).
*
*/
const listSharedProjects = (options) => (options?.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/shared",
	...options
});
/**
* Delete project
*
* Deletes the specified project and all its endpoints, branches, databases, and users.
* Deleted projects can be recovered within 7 days using `POST /projects/{project_id}/recover`.
* To list recoverable projects, use `GET /projects?recoverable=true`.
*
*/
const deleteProject = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}",
	...options
});
/**
* Retrieve project details
*
* Retrieves information about the specified project.
* Returned details include the project settings, compute configuration, history retention, owner information, and current usage metrics.
*
*/
const getProject = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}",
	...options
});
/**
* Update project
*
* Updates the specified project.
* Configurable properties include the project name, default compute settings, history retention period, and IP allowlist.
*
*/
const updateProject = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Recover a deleted project
*
* Recovers a deleted project within the 7-day deletion recovery period.
* Restores branches, endpoints, settings, and connection strings.
* Some integrations require manual reconfiguration after recovery.
* To list recoverable projects, use `GET /projects?recoverable=true`.
*
*/
const recoverProject = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/recover",
	...options
});
/**
* List operations
*
* Retrieves a list of operations for the specified Neon project.
* The number of operations returned can be large.
* To paginate the response, issue an initial request with a `limit` value.
* Then, add the `cursor` value that was returned in the response to the next request.
* Operations older than 6 months may be deleted from our systems.
* If you need more history than that, you should store your own history.
*
*/
const listProjectOperations = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/operations",
	...options
});
/**
* List project access
*
* Retrieves details about users who have access to the project, including the permission `id`, the granted-to email address, and the date project access was granted.
*/
const listProjectPermissions = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/permissions",
	...options
});
/**
* Grant project access
*
* Grants project access to the account associated with the specified email address.
*
*/
const grantPermissionToProject = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/permissions",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Revoke project access
*
* Revokes project access from the user associated with the specified permission `id`. You can retrieve a user's permission `id` by listing project access.
*/
const revokePermissionFromProject = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/permissions/{permission_id}",
	...options
});
/**
* List available shared preload libraries
*
* Returns the shared preload libraries available for the specified project's Postgres version.
* Shared preload libraries are Postgres extensions that require the `shared_preload_libraries`
* setting and a compute restart to activate.
* Use this list to determine which libraries can be enabled in the project's
* `settings.preload_libraries` configuration.
*
*/
const getAvailablePreloadLibraries = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/available_preload_libraries",
	...options
});
/**
* Create a project transfer request
*
* Creates a transfer request for the specified project. The request expires after a set period.
* To accept the request, the recipient calls `PUT /projects/{project_id}/transfer_requests/{request_id}`
* or uses the Neon Console claim link.
* The optional `ru` parameter redirects the recipient after acceptance.
*
*/
const createProjectTransferRequest = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/transfer_requests",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Accept a project transfer request
*
* Accepts a transfer request for the specified project, transferring it to the specified organization
* or user. If org_id is not passed, the project will be transferred to the current user or organization account.
*
*/
const acceptProjectTransferRequest = (options) => (options.client ?? client).put({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/transfer_requests/{request_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List JWKS URLs
*
* Returns the JWKS URLs available for verifying JWTs used as the authentication mechanism for the specified project.
*
*/
const getProjectJwks = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/jwks",
	...options
});
/**
* Add JWKS URL
*
* Adds a JWKS URL to the specified project for verifying JWTs used as the authentication mechanism.
*
* The URL must be a valid HTTPS URL that returns a JSON Web Key Set.
*
* The `provider_name` field allows you to specify which authentication provider you're using (e.g., Clerk, Auth0, AWS Cognito).
*
* The `branch_id` scopes the JWKS URL to specific branches; if not specified, it applies to all branches.
*
* The `role_names` scopes the URL to specific roles; if not specified, default roles are used (`authenticator`, `authenticated`, `anonymous`).
*
* The `jwt_audience` specifies which `aud` values are accepted in JWTs.
*
*/
const addProjectJwks = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/jwks",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete JWKS URL
*
* Removes the specified JWKS URL from the project.
* JWTs signed by keys from the removed URL can no longer authenticate to the project's endpoints.
*
*/
const deleteProjectJwks = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/jwks/{jwks_id}",
	...options
});
/**
* Delete Neon Data API
*
* Deletes the Neon Data API for the specified branch.
* Existing connections using the Data API endpoint will fail after deletion.
*
*/
const deleteProjectBranchDataApi = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/data-api/{database_name}",
	...options
});
/**
* Retrieve Neon Data API configuration
*
* Retrieves the Neon Data API configuration for the specified branch,
* including endpoint URL, enabled state, and database settings.
*
*/
const getProjectBranchDataApi = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/data-api/{database_name}",
	...options
});
/**
* Update Neon Data API
*
* Updates the Neon Data API configuration for the specified branch.
* You can optionally provide settings to update the Data API configuration.
* The schema cache is always refreshed as part of this operation.
*
*/
const updateProjectBranchDataApi = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/data-api/{database_name}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Create Neon Data API
*
* Creates a new instance of Neon Data API in the specified branch.
* The Data API exposes a REST interface over the branch database. The `database_name` path parameter determines which database the API serves.
*
*/
const createProjectBranchDataApi = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/data-api/{database_name}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Create Neon Auth integration
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth` instead. Creates a project on a third-party authentication provider's platform for use with Neon Auth.
* Use this endpoint if the frontend integration flow can't be used.
*
*
* @deprecated
*/
const createNeonAuthIntegration = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/auth/create",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Disable Neon Auth for the branch
*
* Disables the Neon Auth integration for the specified branch, removing the connection
* to the authentication provider.
* If `delete_data` is `true`, also deletes the `neon_auth` schema and all associated tables
* from the branch database.
* The integration can be re-enabled by calling `POST /projects/{project_id}/branches/{branch_id}/auth`.
*
*/
const disableNeonAuth = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve Neon Auth details for the branch
*
* Retrieves the Neon Auth integration details for the specified branch,
* including the auth provider type and integration status.
*
*/
const getNeonAuth = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth",
	...options
});
/**
* Enable Neon Auth for the branch
*
* Enables Neon Auth for the specified branch by connecting it to an authentication provider.
* Creating the integration provisions the `neon_auth` schema in the branch database, which stores user identity data synchronized from the provider.
*
*/
const createNeonAuth = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete trusted redirect URI domain
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/domains` instead. Deletes a domain from the redirect_uri whitelist for the specified project.
*
*
* @deprecated
*/
const deleteNeonAuthDomainFromRedirectUriWhitelist = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/domains",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List trusted redirect URI domains
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/domains` instead. Lists the domains in the redirect_uri whitelist for the specified project.
*
*
* @deprecated
*/
const listNeonAuthRedirectUriWhitelistDomains = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/domains",
	...options
});
/**
* Add trusted redirect URI domain
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/domains` instead. Adds a domain to the redirect_uri whitelist for the specified project.
*
*
* @deprecated
*/
const addNeonAuthDomainToRedirectUriWhitelist = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/domains",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete domain from redirect_uri whitelist
*
* Removes a domain from the redirect URI whitelist for the specified branch.
* After removal, the domain can no longer be used as a redirect target after authentication.
*
*/
const deleteBranchNeonAuthTrustedDomain = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/domains",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List domains in redirect_uri whitelist
*
* Lists the trusted domains in the redirect URI whitelist for the specified branch.
* Only domains in this list are permitted as redirect targets after authentication.
*
*/
const listBranchNeonAuthTrustedDomains = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/domains",
	...options
});
/**
* Add domain to redirect_uri whitelist
*
* Adds a domain to the redirect URI whitelist for the specified branch.
* Only domains in this list are permitted as redirect targets after authentication.
*
*/
const addBranchNeonAuthTrustedDomain = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/domains",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Create Auth Provider SDK keys
*
* Generates SDK or API Keys for the auth provider. These might be called different things depending
* on the auth provider you're using, but are generally used for setting up the frontend and backend SDKs.
*
*/
const createNeonAuthProviderSdkKeys = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/auth/keys",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Create new auth user
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/users` instead. Creates a new user in Neon Auth.
* The user will be created in your neon_auth.users_sync table and automatically propagated to your auth project, whether Neon-managed or provider-owned.
*
*
* @deprecated
*/
const createNeonAuthNewUser = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/auth/user",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Create new auth user
*
* Creates a new user in the Neon Auth user directory for the specified branch.
* The user is created in the `neon_auth.users_sync` table and can immediately authenticate
* using the branch's configured auth providers.
*
*/
const createBranchNeonAuthNewUser = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/users",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete auth user
*
* Deletes the specified user from the Neon Auth user directory for the specified branch.
* Removes the user record from `neon_auth.users_sync`. This action cannot be undone.
*
*/
const deleteBranchNeonAuthUser = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/users/{auth_user_id}",
	...options
});
/**
* Update auth user role
*
* Updates the role of a user in the Neon Auth user directory for the specified branch.
* The role controls the user's level of access within the Neon Auth integration.
*
*/
const updateNeonAuthUserRole = (options) => (options.client ?? client).put({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/users/{auth_user_id}/role",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete auth user
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/users/{auth_user_id}` instead. Deletes the auth user for the specified project.
*
*
* @deprecated
*/
const deleteNeonAuthUser = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/users/{auth_user_id}",
	...options
});
/**
* Transfer Neon-managed auth project to your own account
*
* Transfers ownership of your Neon-managed auth project to your own auth provider account.
*
*/
const transferNeonAuthProviderProject = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/auth/transfer_ownership",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List active integrations with auth providers
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth` instead.
*
* @deprecated
*/
const listNeonAuthIntegrations = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/integrations",
	...options
});
/**
* List OAuth providers
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/oauth_providers` instead. Lists the OAuth providers for the specified project.
*
*
* @deprecated
*/
const listNeonAuthOauthProviders = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/oauth_providers",
	...options
});
/**
* Add an OAuth provider
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/oauth_providers` instead.
* Adds an OAuth provider to the specified project.
*
*
* @deprecated
*/
const addNeonAuthOauthProvider = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/oauth_providers",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List OAuth providers for the branch
*
* Lists the OAuth providers configured for the specified branch's Neon Auth integration.
*
*/
const listBranchNeonAuthOauthProviders = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/oauth_providers",
	...options
});
/**
* Add an OAuth provider
*
* Adds an OAuth provider configuration to the specified branch's Neon Auth integration.
* After adding, users can authenticate using the configured provider.
*
*/
const addBranchNeonAuthOauthProvider = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/oauth_providers",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete OAuth provider
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/oauth_providers/{oauth_provider_id}` instead. Deletes a OAuth provider from the specified project.
*
*
* @deprecated
*/
const deleteNeonAuthOauthProvider = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/oauth_providers/{oauth_provider_id}",
	...options
});
/**
* Update OAuth provider
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/oauth_providers/{oauth_provider_id}` instead. Updates a OAuth provider for the specified project.
*
*
* @deprecated
*/
const updateNeonAuthOauthProvider = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/oauth_providers/{oauth_provider_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete OAuth provider
*
* Deletes a OAuth provider from the specified project.
*
*/
const deleteBranchNeonAuthOauthProvider = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/oauth_providers/{oauth_provider_id}",
	...options
});
/**
* Update OAuth provider
*
* Updates a OAuth provider for the specified project.
*
*/
const updateBranchNeonAuthOauthProvider = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/oauth_providers/{oauth_provider_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve email server configuration
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/email_provider` instead. Gets the email server configuration for the specified project.
*
*
* @deprecated
*/
const getNeonAuthEmailServer = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/email_server",
	...options
});
/**
* Update email server configuration
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth/email_provider` instead. Updates the email server configuration for the specified project.
*
*
* @deprecated
*/
const updateNeonAuthEmailServer = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/email_server",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Send test email
*
* Sends a test email using the configured email server settings to verify SMTP connectivity and credentials.
* The request body must include the SMTP server settings
* (`host`, `port`, `username`, `password`, `sender_email`, `sender_name`) and the `recipient_email` address.
*
*/
const sendNeonAuthTestEmail = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/send_test_email",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve email and password configuration
*
* Retrieves the email and password authentication configuration for the specified branch's Neon Auth integration,
* including whether it is enabled and the email verification method.
*
*/
const getNeonAuthEmailAndPasswordConfig = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/email_and_password",
	...options
});
/**
* Update email and password configuration
*
* Updates the email and password authentication configuration for the specified branch's Neon Auth integration.
* Only the fields provided in the request body are updated.
*
*/
const updateNeonAuthEmailAndPasswordConfig = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/email_and_password",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve email provider configuration
*
* Retrieves the email provider configuration for the specified branch's Neon Auth integration,
* including the provider type and server settings.
*
*/
const getNeonAuthEmailProvider = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/email_provider",
	...options
});
/**
* Update email provider configuration
*
* Updates the email provider configuration for the specified branch's Neon Auth integration.
* The email provider handles transactional messages such as verification emails and password reset links.
*
*/
const updateNeonAuthEmailProvider = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/email_provider",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete integration with auth provider
*
* DEPRECATED, use `/projects/{project_id}/branches/{branch_id}/auth` instead.
*
* @deprecated
*/
const deleteNeonAuthIntegration = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/auth/integration/{auth_provider}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve connection URI
*
* Retrieves a connection URI for the specified database.
* The URI uses the standard PostgreSQL connection string format. Set `pooled=true` to include the `-pooler` suffix for a connection pooler URI.
*
*/
const getConnectionUri = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/connection_uri",
	...options
});
/**
* Retrieve localhost allow setting
*
* Retrieves the localhost allow setting for the specified branch's Neon Auth integration.
* When enabled, authentication flows work from `localhost` without adding it to the redirect URI whitelist.
*
*/
const getNeonAuthAllowLocalhost = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/allow_localhost",
	...options
});
/**
* Update localhost allow setting
*
* Updates the localhost allow setting for the specified branch's Neon Auth integration.
* When enabled, authentication flows work from `localhost` without adding it to the redirect URI whitelist.
*
*/
const updateNeonAuthAllowLocalhost = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/allow_localhost",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve Neon Auth plugin configurations
*
* Returns all plugin configurations for Neon Auth in a single response.
* This endpoint aggregates organization, email provider, email and password,
* OAuth providers, and localhost settings.
*
*/
const getNeonAuthPluginConfigs = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/plugins",
	...options
});
/**
* Update organization plugin configuration
*
* Updates the organization plugin configuration for Neon Auth.
* The organization plugin enables multi-tenant organization support.
*
*/
const updateNeonAuthOrganizationPlugin = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/plugins/organization",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Update auth configuration
*
* Updates the auth configuration for the branch.
* Currently supports updating the application name used in auth emails.
*
*/
const updateNeonAuthConfig = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/config",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Update magic link plugin configuration
*
* Updates the magic link plugin configuration for Neon Auth.
* The magic link plugin enables passwordless authentication via email magic links.
*
*/
const updateNeonAuthMagicLinkPlugin = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/plugins/magic-link",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve phone number plugin configuration
*
* Returns the phone number plugin configuration for Neon Auth.
* The phone number plugin enables phone-based OTP authentication.
*
*/
const getNeonAuthPhoneNumberPlugin = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/plugins/phone-number",
	...options
});
/**
* Update phone number plugin configuration
*
* Updates the phone number plugin configuration for Neon Auth.
* Only the fields provided in the request body are updated; omitted fields retain their current values.
* The phone number plugin enables phone-based OTP authentication.
* OTP codes are delivered via the `send.otp` webhook event with `delivery_preference: "sms"`.
* A webhook must be configured with the `send.otp` event enabled for SMS delivery to work.
*
*/
const updateNeonAuthPhoneNumberPlugin = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/plugins/phone-number",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve Neon Auth webhook configuration
*
* Returns the webhook configuration for the specified branch's Neon Auth integration,
* including the endpoint URL and the events that trigger it.
*
*/
const getNeonAuthWebhookConfig = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/webhooks",
	...options
});
/**
* Update Neon Auth webhook configuration
*
* Updates the webhook configuration for the specified branch's Neon Auth integration.
* Webhooks notify an external endpoint when auth events occur, such as user creation or sign-in.
*
*/
const updateNeonAuthWebhookConfig = (options) => (options.client ?? client).put({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/auth/webhooks",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List branches
*
* Retrieves a list of branches for the specified project.
*
* Each Neon project has a root branch named `main`.
* A `branch_id` value has a `br-` prefix.
* A project may contain child branches that were branched from `main` or from another branch.
* A parent branch is identified by the `parent_id` value, which is the `id` of the parent branch.
* For related information, see [Manage branches](https://neon.com/docs/manage/branches/).
*
*/
const listProjectBranches = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches",
	...options
});
/**
* Create branch
*
* Creates a branch in the specified project.
* No request body is required, but you can specify one to create a compute endpoint or select a non-default parent branch.
* By default, the branch is created from the project's default branch with no compute endpoint, and the branch name is auto-generated.
* To access the branch, add a `read_write` endpoint.
* Each branch supports one read-write endpoint and multiple read-only endpoints.
* For related information, see [Manage branches](https://neon.com/docs/manage/branches/).
*
*/
const createProjectBranch = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Create anonymized branch
*
* Creates a new branch with anonymized data using PostgreSQL Anonymizer for static masking.
* This allows developers to work with masked production data.
* Optionally, provide `masking_rules` to set initial masking rules for the branch
* and `start_anonymization` to automatically start anonymization after creation. This
* combines functionality of updating masking rules and starting anonymization into the
* branch creation request.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const createProjectBranchAnonymized = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branch_anonymized",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve number of branches
*
* Retrieves the total number of branches in the specified project.
* Supports an optional `search` parameter to count branches matching a name filter.
*
*/
const countProjectBranches = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/count",
	...options
});
/**
* Delete branch
*
* Deletes the specified branch from a project and places all compute endpoints into an idle state, breaking existing client connections.
*
* The deletion completes after all operations finish.
* You cannot delete a project's root or default branch, or a branch that has a child branch.
* A project must have at least one branch.
*
* By default, deleted branches can be recovered within a 7-day grace period.
* Use the `hard_delete` parameter to permanently delete the branch immediately.
* For related information, see [Manage branches](https://neon.com/docs/manage/branches/).
*
*/
const deleteProjectBranch = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}",
	...options
});
/**
* Retrieve branch details
*
* Retrieves information about the specified branch.
* A `branch_id` value has a `br-` prefix.
*
* Each Neon project is initially created with a root and default branch named `main`.
* A project can contain one or more branches.
* A parent branch is identified by a `parent_id` value, which is the `id` of the parent branch.
* For related information, see [Manage branches](https://neon.com/docs/manage/branches/).
*
*/
const getProjectBranch = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}",
	...options
});
/**
* Update branch
*
* Updates the specified branch.
* For more information, see [Manage branches](https://neon.com/docs/manage/branches/).
*
*/
const updateProjectBranch = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Restore branch to a historical state
*
* Restores a branch to an earlier state in its own or another branch's history
* by specifying an LSN or timestamp.
* Creates a new branch from the historical state.
*
*/
const restoreProjectBranch = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/restore",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve database schema
*
* Retrieves the schema from the specified database. The `lsn` and `timestamp` values cannot be specified at the same time. If both are omitted, the database schema is retrieved from database's head.
*/
const getProjectBranchSchema = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/schema",
	...options
});
/**
* Compare database schema
*
* Compares the schema from the specified database with another branch's schema.
*/
const getProjectBranchSchemaComparison = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/compare_schema",
	...options
});
/**
* Retrieve masking rules
*
* Retrieves the masking rules for the specified anonymized branch.
* Masking rules define how sensitive data should be anonymized using PostgreSQL Anonymizer.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const getMaskingRules = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/masking_rules",
	...options
});
/**
* Update masking rules
*
* Updates the masking rules for the specified anonymized branch.
* Masking rules define how sensitive data should be anonymized using PostgreSQL Anonymizer.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const updateMaskingRules = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/masking_rules",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve anonymized branch status
*
* Retrieves the current status of an anonymized branch, including its state and progress information.
* This endpoint allows you to monitor the anonymization process from initialization through completion.
* Only anonymized branches will have status information available.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const getAnonymizedBranchStatus = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/anonymized_status",
	...options
});
/**
* Start anonymization
*
* Starts the anonymization process for an anonymized branch that is in the initialized, error, or anonymized state.
* This will apply all defined masking rules to anonymize sensitive data in the branch databases.
* The branch must be an anonymized branch to start anonymization.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const startAnonymization = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/anonymize",
	...options
});
/**
* Set branch as default
*
* Sets the specified branch as the project's default branch.
* The default designation is automatically removed from the previous default branch.
* For more information, see [Manage branches](https://neon.com/docs/manage/branches/).
*
*/
const setDefaultProjectBranch = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/set_as_default",
	...options
});
/**
* Finalize branch restore from snapshot
*
* Finalize the restore operation for a branch created from a snapshot.
* This operation updates the branch so it functions as the original branch it replaced.
* This includes:
* - Reassigning any computes from the original branch to the restored branch (this will restart the computes)
* - Renaming the restored branch to the original branch's name
* - Renaming the original branch so it no longer uses the original name
*
* This operation only applies to branches created using the `restoreSnapshot` endpoint with `finalize_restore: false`.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const finalizeRestoreBranch = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/finalize_restore",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List branch endpoints
*
* Retrieves a list of compute endpoints for the specified branch.
* Neon permits only one read-write compute endpoint per branch.
* A branch can have multiple read-only compute endpoints.
*
*/
const listProjectBranchEndpoints = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/endpoints",
	...options
});
/**
* List databases
*
* Retrieves a list of databases for the specified branch.
* A branch can have multiple databases.
* For related information, see [Manage databases](https://neon.com/docs/manage/databases/).
*
*/
const listProjectBranchDatabases = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/databases",
	...options
});
/**
* Create database
*
* Creates a database in the specified branch.
* A branch can have multiple databases.
* For related information, see [Manage databases](https://neon.com/docs/manage/databases/).
*
*/
const createProjectBranchDatabase = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/databases",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete database
*
* Deletes the specified database from the branch.
* For related information, see [Manage databases](https://neon.com/docs/manage/databases/).
*
*/
const deleteProjectBranchDatabase = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/databases/{database_name}",
	...options
});
/**
* Retrieve database details
*
* Retrieves information about the specified database.
* For related information, see [Manage databases](https://neon.com/docs/manage/databases/).
*
*/
const getProjectBranchDatabase = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/databases/{database_name}",
	...options
});
/**
* Update database
*
* Updates the specified database in the branch.
* For related information, see [Manage databases](https://neon.com/docs/manage/databases/).
*
*/
const updateProjectBranchDatabase = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/databases/{database_name}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List roles
*
* Retrieves a list of Postgres roles from the specified branch.
* For related information, see [Manage roles](https://neon.com/docs/manage/roles/).
*
*/
const listProjectBranchRoles = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/roles",
	...options
});
/**
* Create role
*
* Creates a Postgres role in the specified branch.
* For related information, see [Manage roles](https://neon.com/docs/manage/roles/).
*
* Connections established to the active compute endpoint will be dropped.
* If the compute endpoint is idle, the endpoint becomes active for a short period of time and is suspended afterward.
*
*/
const createProjectBranchRole = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/roles",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete role
*
* Deletes the specified Postgres role from the branch.
* For related information, see [Manage roles](https://neon.com/docs/manage/roles/).
*
*/
const deleteProjectBranchRole = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/roles/{role_name}",
	...options
});
/**
* Retrieve role details
*
* Retrieves details about the specified role.
* In Neon, the terms "role" and "user" are synonymous.
* For related information, see [Manage roles](https://neon.com/docs/manage/roles/).
*
*/
const getProjectBranchRole = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/roles/{role_name}",
	...options
});
/**
* Retrieve role password
*
* Retrieves the password for the specified Postgres role, if possible.
* For related information, see [Manage roles](https://neon.com/docs/manage/roles/).
*
*/
const getProjectBranchRolePassword = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/roles/{role_name}/reveal_password",
	...options
});
/**
* Reset role password
*
* Resets the password for the specified Postgres role.
* Returns a new password and operations. The new password is ready to use when the last operation finishes.
* The old password remains valid until last operation finishes.
* Connections to the compute endpoint are dropped. If idle,
* the compute endpoint becomes active for a short period of time.
*
* For related information, see [Manage roles](https://neon.com/docs/manage/roles/).
*
*/
const resetProjectBranchRolePassword = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/roles/{role_name}/reset_password",
	...options
});
/**
* List VPC endpoint restrictions
*
* Lists VPC endpoint restrictions for the specified Neon project.
*
*/
const listProjectVpcEndpoints = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/vpc_endpoints",
	...options
});
/**
* Delete VPC endpoint restriction
*
* Removes the specified VPC endpoint restriction from a Neon project.
*
*/
const deleteProjectVpcEndpoint = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/vpc_endpoints/{vpc_endpoint_id}",
	...options
});
/**
* Set VPC endpoint restriction
*
* Sets or updates a VPC endpoint restriction for a Neon project.
* When a VPC endpoint restriction is set, the project only accepts connections
* from the specified VPC.
* A VPC endpoint can be set as a restriction only after it is assigned to the
* parent organization of the Neon project.
*
*/
const assignProjectVpcEndpoint = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/vpc_endpoints/{vpc_endpoint_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List compute endpoints
*
* Retrieves a list of compute endpoints for the specified project.
* A compute endpoint is a Neon compute instance.
* For information about compute endpoints, see [Manage computes](https://neon.com/docs/manage/endpoints/).
*
*/
const listProjectEndpoints = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/endpoints",
	...options
});
/**
* Create compute endpoint
*
* Creates a compute endpoint for the specified branch.
* A compute endpoint is a Neon compute instance.
* There is a maximum of one read-write compute endpoint per branch.
* If the specified branch already has a read-write compute endpoint, the operation fails.
* A branch can have multiple read-only compute endpoints.
*
* For more information about compute endpoints, see [Manage computes](https://neon.com/docs/manage/endpoints/).
*
*/
const createProjectEndpoint = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/endpoints",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Delete compute endpoint
*
* Deletes the specified compute endpoint.
* A compute endpoint is a Neon compute instance.
* Deleting a compute endpoint drops existing network connections to the compute endpoint.
* The deletion is completed when the last operation in the chain finishes successfully.
*
* An `endpoint_id` has an `ep-` prefix.
* For information about compute endpoints, see [Manage computes](https://neon.com/docs/manage/endpoints/).
*
*/
const deleteProjectEndpoint = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/endpoints/{endpoint_id}",
	...options
});
/**
* Retrieve compute endpoint details
*
* Retrieves information about the specified compute endpoint.
* A compute endpoint is a Neon compute instance.
* An `endpoint_id` has an `ep-` prefix.
* For information about compute endpoints, see [Manage computes](https://neon.com/docs/manage/endpoints/).
*
*/
const getProjectEndpoint = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/endpoints/{endpoint_id}",
	...options
});
/**
* Update compute endpoint
*
* Updates the specified compute endpoint.
*
* An `endpoint_id` has an `ep-` prefix. A `branch_id` has a `br-` prefix.
* For more information about compute endpoints, see [Manage computes](https://neon.com/docs/manage/endpoints/).
*
* If the returned list of operations is not empty, the compute endpoint is not ready to use.
* The client must wait for the last operation to finish before using the compute endpoint.
* If the compute endpoint was idle before the update, it becomes active for a short period of time,
* and the control plane suspends it again after the update.
*
*/
const updateProjectEndpoint = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/endpoints/{endpoint_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Start compute endpoint
*
* Starts a compute endpoint.
* The compute endpoint is ready to use after the last operation in the chain finishes successfully.
*
* An `endpoint_id` has an `ep-` prefix.
* For information about compute endpoints, see [Manage computes](https://neon.com/docs/manage/endpoints/).
*
*/
const startProjectEndpoint = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/endpoints/{endpoint_id}/start",
	...options
});
/**
* Suspend compute endpoint
*
* Suspends the specified compute endpoint.
* An `endpoint_id` has an `ep-` prefix.
* For information about compute endpoints, see [Manage computes](https://neon.com/docs/manage/endpoints/).
*
*/
const suspendProjectEndpoint = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/endpoints/{endpoint_id}/suspend",
	...options
});
/**
* Restart compute endpoint
*
* Restarts the specified compute endpoint by immediately suspending it and then starting it again.
* An `endpoint_id` has an `ep-` prefix.
* For information about compute endpoints, see [Manage computes](https://neon.com/docs/manage/endpoints/).
*
*/
const restartProjectEndpoint = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/endpoints/{endpoint_id}/restart",
	...options
});
/**
* Retrieve project consumption metrics (legacy plans)
*
* Retrieves consumption metrics for Scale, Business, and Enterprise plan projects. History begins at the time of upgrade.
* Results are ordered by time in ascending order (oldest to newest).
* Issuing a call to this API does not wake a project's compute endpoint.
*
*/
const getConsumptionHistoryPerProject = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/consumption_history/projects",
	...options
});
/**
* Retrieve project consumption metrics
*
* Returns consumption metrics for up to `limit` projects per page. If `project_ids` is omitted,
* projects in the organization are included across pages (use `cursor`). If `project_ids` is
* provided, the response is limited to those projects (up to 100). Available for accounts on
* Launch, Scale, Agent, Business, and Enterprise plans.
*
* History starts when the account upgrades to an eligible plan.
*
* The `metrics` query parameter is required. Supported values:
* `compute_unit_seconds`, `root_branch_bytes_month`, `child_branch_bytes_month`,
* `instant_restore_bytes_month`, `public_network_transfer_bytes`, `private_network_transfer_bytes`,
* `extra_branches_month`, `snapshot_storage_bytes_month`.
*
* Consumption metrics within each project are returned in ascending time order (oldest first).
* This request does not wake project computes.
*
*/
const getConsumptionHistoryPerProjectV2 = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/consumption_history/v2/projects",
	...options
});
/**
* Retrieve branch consumption metrics
*
* Returns consumption metrics for each branch across one or more projects listed in
* `project_ids` (1 to 100 projects). Available for accounts on paid usage-based Launch, Scale,
* Agent, and Enterprise plans.
*
* History starts when the account first ingests branch-level consumption data.
*
* The `metrics` query parameter is required. Only these six values are supported on this
* endpoint:
* `compute_unit_seconds`, `root_branch_bytes_month`, `child_branch_bytes_month`,
* `instant_restore_bytes_month`, `public_network_transfer_bytes`, `private_network_transfer_bytes`.
*
* This endpoint does not support `extra_branches_month` or `snapshot_storage_bytes_month`.
* Use `GET /consumption_history/v2/projects` for those.
*
* Consumption metrics within each branch are returned in ascending time order (oldest first).
* This request does not wake project computes.
*
*/
const getConsumptionHistoryPerBranchV2 = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/consumption_history/v2/branches",
	...options
});
/**
* Retrieve organization details
*
* Retrieves details for the specified organization, including its name, plan, and configuration.
*
*/
const getOrganization = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}",
	...options
});
/**
* List organization API keys
*
* Retrieves the API keys for the specified organization.
* The response does not include API key tokens. A token is only provided when creating an API key.
* API keys can also be managed in the Neon Console.
* For more information, see [Manage API keys](https://neon.com/docs/manage/api-keys/).
*
*/
const listOrgApiKeys = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/api_keys",
	...options
});
/**
* Create organization API key
*
* Creates an API key for the specified organization.
* The `key_name` is a user-specified name for the key.
* Returns an `id` and `key`; the `key` is a randomly generated, 64-bit token required to access the Neon API.
* Store the key securely — it is only returned once.
* API keys can also be managed in the Neon Console.
* See [Manage API keys](https://neon.com/docs/manage/api-keys/).
*
*/
const createOrgApiKey = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/api_keys",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Revoke organization API key
*
* Revokes the specified organization API key.
* An API key that is no longer needed can be revoked.
* This action cannot be reversed.
* API keys can also be managed in the Neon Console.
* See [Manage API keys](https://neon.com/docs/manage/api-keys/).
*
*/
const revokeOrgApiKey = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/api_keys/{key_id}",
	...options
});
/**
* Remove organization spending limit
*
* Removes the configured monthly spending limit for the specified organization.
* Idempotent — removing an already-unset limit still succeeds.
* Available to organization admins on Launch and Scale plans only.
*
*/
const deleteOrganizationSpendingLimit = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/billing/spending_limit",
	...options
});
/**
* Retrieve organization spending limit
*
* Returns the configured monthly spending limit for the specified organization.
* `spending_limit_cents: null` indicates that no limit is currently set.
* Available to organization members with read access on Launch and Scale plans only.
*
*/
const getOrganizationSpendingLimit = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/billing/spending_limit",
	...options
});
/**
* Set organization spending limit
*
* Sets the monthly spending limit for the specified organization.
* To remove a previously configured limit, send a DELETE request to this endpoint.
* When a limit is configured, email notifications are sent at 80% and 100% of the limit.
* Computes are not suspended when the limit is reached.
* Available to organization admins on Launch and Scale plans only.
*
*/
const setOrganizationSpendingLimit = (options) => (options.client ?? client).put({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/billing/spending_limit",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List organization members
*
* Retrieves a paginated list of members for the specified organization.
*
*/
const getOrganizationMembers = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/members",
	...options
});
/**
* Remove organization member
*
* Removes the specified member from the organization.
* Only organization admins can perform this action.
* The last admin in an organization cannot be removed.
*
*/
const removeOrganizationMember = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/members/{member_id}",
	...options
});
/**
* Retrieve organization member details
*
* Retrieves information about the specified organization member.
*
*/
const getOrganizationMember = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/members/{member_id}",
	...options
});
/**
* Update role for organization member
*
* Updates the role of an existing member in the specified organization.
* The requested role must be valid for the organization.
* Only organization admins can call this endpoint.
*
*/
const updateOrganizationMember = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/members/{member_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List organization invitations
*
* Retrieves pending and accepted invitations for the specified organization.
*
*/
const getOrganizationInvitations = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/invitations",
	...options
});
/**
* Create organization invitations
*
* Creates invitations for a specific organization.
* If the invited user has an existing account, they automatically join as a member.
* If they don't yet have an account, they are invited to create one, after which they become a member.
* Each invited user receives an email notification.
*
*/
const createOrganizationInvitations = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/invitations",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Transfer projects between organizations
*
* Transfers selected projects, identified by their IDs, from your organization to another specified organization.
*
*/
const transferProjectsFromOrgToOrg = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{source_org_id}/projects/transfer",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List VPC endpoints across all regions
*
* Retrieves the list of VPC endpoints for the specified Neon organization across all regions.
*
*/
const listOrganizationVpcEndpointsAllRegions = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/vpc/vpc_endpoints",
	...options
});
/**
* List VPC endpoints
*
* Retrieves the list of VPC endpoints for the specified Neon organization.
*
*/
const listOrganizationVpcEndpoints = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/vpc/region/{region_id}/vpc_endpoints",
	...options
});
/**
* Delete VPC endpoint
*
* Deletes the VPC endpoint from the specified Neon organization.
* If you delete a VPC endpoint from a Neon organization, that VPC endpoint cannot
* be added back to the Neon organization.
*
*/
const deleteOrganizationVpcEndpoint = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/vpc/region/{region_id}/vpc_endpoints/{vpc_endpoint_id}",
	...options
});
/**
* Retrieve VPC endpoint details
*
* Retrieves the current state and configuration details of a specified VPC endpoint.
*
*/
const getOrganizationVpcEndpointDetails = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/vpc/region/{region_id}/vpc_endpoints/{vpc_endpoint_id}",
	...options
});
/**
* Assign or update VPC endpoint
*
* Assigns a VPC endpoint to a Neon organization or updates its existing assignment.
*
*/
const assignOrganizationVpcEndpoint = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/organizations/{org_id}/vpc/region/{region_id}/vpc_endpoints/{vpc_endpoint_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* List supported regions
*
* Lists supported Neon regions.
*
* **Note:** Not all regions are available to all organizations. Pass the `org_id`
* parameter to get an accurate list of regions available to your organization.
*
*/
const getActiveRegions = (options) => (options?.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/regions",
	...options
});
/**
* Retrieve current user details
*
* Retrieves information about the currently authenticated Neon user,
* including account identifiers, plan details, and linked auth accounts.
*
*/
const getCurrentUserInfo = (options) => (options?.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/users/me",
	...options
});
/**
* List organizations for the current user
*
* Retrieves the organizations that the currently authenticated user belongs to.
*
*/
const getCurrentUserOrganizations = (options) => (options?.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/users/me/organizations",
	...options
});
/**
* Transfer projects from personal account to organization
*
* DEPRECATED. Personal accounts have been migrated to organizations, making this operation no longer applicable.
*
*
* @deprecated
*/
const transferProjectsFromUserToOrg = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/users/me/projects/transfer",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve request authentication details
*
* Returns authentication details for the credentials used in the request,
* including the credential type (API key, Bearer token, or OAuth session)
* and the associated identity.
*
*/
const getAuthDetails = (options) => (options?.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/auth",
	...options
});
/**
* Create snapshot
*
* Creates a snapshot from the specified branch.
* This operation may initiate an asynchronous process.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const createSnapshot = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/snapshot",
	...options
});
/**
* List project snapshots
*
* Lists the snapshots for the specified project.
* Each snapshot represents a point-in-time backup of the project data.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const listSnapshots = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/snapshots",
	...options
});
/**
* Delete snapshot
*
* Deletes the specified snapshot.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const deleteSnapshot = (options) => (options.client ?? client).delete({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/snapshots/{snapshot_id}",
	...options
});
/**
* Update snapshot
*
* Updates the specified snapshot.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const updateSnapshot = (options) => (options.client ?? client).patch({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/snapshots/{snapshot_id}",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Restore snapshot
*
* Restores the specified snapshot to a new branch,
* and optionally finalizes the restore operation to replace the original branch.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const restoreSnapshot = (options) => (options.client ?? client).post({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/snapshots/{snapshot_id}/restore",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
/**
* Retrieve backup schedule
*
* Returns the backup schedule for the specified branch, including the configured snapshot frequencies.
*
* **Note**: This endpoint is currently in Beta.
*
*/
const getSnapshotSchedule = (options) => (options.client ?? client).get({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/backup_schedule",
	...options
});
/**
* Update backup schedule
*
* Updates the backup schedule for the specified branch.
* The schedule defines how often automatic snapshots are created (e.g., `hourly`, `daily`).
*
* **Note**: This endpoint is currently in Beta.
*
*/
const setSnapshotSchedule = (options) => (options.client ?? client).put({
	security: [
		{
			scheme: "bearer",
			type: "http"
		},
		{
			in: "cookie",
			name: "zenith",
			type: "apiKey"
		},
		{
			in: "cookie",
			name: "keycloak_token",
			type: "apiKey"
		}
	],
	url: "/projects/{project_id}/branches/{branch_id}/backup_schedule",
	...options,
	headers: {
		"Content-Type": "application/json",
		...options.headers
	}
});
//#endregion
export { acceptProjectTransferRequest, addBranchNeonAuthOauthProvider, addBranchNeonAuthTrustedDomain, addNeonAuthDomainToRedirectUriWhitelist, addNeonAuthOauthProvider, addProjectJwks, assignOrganizationVpcEndpoint, assignProjectVpcEndpoint, countProjectBranches, createApiKey, createBranchNeonAuthNewUser, createNeonAuth, createNeonAuthIntegration, createNeonAuthNewUser, createNeonAuthProviderSdkKeys, createOrgApiKey, createOrganizationInvitations, createProject, createProjectBranch, createProjectBranchAnonymized, createProjectBranchDataApi, createProjectBranchDatabase, createProjectBranchRole, createProjectEndpoint, createProjectTransferRequest, createSnapshot, deleteBranchNeonAuthOauthProvider, deleteBranchNeonAuthTrustedDomain, deleteBranchNeonAuthUser, deleteNeonAuthDomainFromRedirectUriWhitelist, deleteNeonAuthIntegration, deleteNeonAuthOauthProvider, deleteNeonAuthUser, deleteOrganizationSpendingLimit, deleteOrganizationVpcEndpoint, deleteProject, deleteProjectBranch, deleteProjectBranchDataApi, deleteProjectBranchDatabase, deleteProjectBranchRole, deleteProjectEndpoint, deleteProjectJwks, deleteProjectVpcEndpoint, deleteSnapshot, disableNeonAuth, finalizeRestoreBranch, getActiveRegions, getAnonymizedBranchStatus, getAuthDetails, getAvailablePreloadLibraries, getConnectionUri, getConsumptionHistoryPerBranchV2, getConsumptionHistoryPerProject, getConsumptionHistoryPerProjectV2, getCurrentUserInfo, getCurrentUserOrganizations, getMaskingRules, getNeonAuth, getNeonAuthAllowLocalhost, getNeonAuthEmailAndPasswordConfig, getNeonAuthEmailProvider, getNeonAuthEmailServer, getNeonAuthPhoneNumberPlugin, getNeonAuthPluginConfigs, getNeonAuthWebhookConfig, getOrganization, getOrganizationInvitations, getOrganizationMember, getOrganizationMembers, getOrganizationSpendingLimit, getOrganizationVpcEndpointDetails, getProject, getProjectAdvisorSecurityIssues, getProjectBranch, getProjectBranchDataApi, getProjectBranchDatabase, getProjectBranchRole, getProjectBranchRolePassword, getProjectBranchSchema, getProjectBranchSchemaComparison, getProjectEndpoint, getProjectJwks, getProjectOperation, getSnapshotSchedule, grantPermissionToProject, listApiKeys, listBranchNeonAuthOauthProviders, listBranchNeonAuthTrustedDomains, listNeonAuthIntegrations, listNeonAuthOauthProviders, listNeonAuthRedirectUriWhitelistDomains, listOrgApiKeys, listOrganizationVpcEndpoints, listOrganizationVpcEndpointsAllRegions, listProjectBranchDatabases, listProjectBranchEndpoints, listProjectBranchRoles, listProjectBranches, listProjectEndpoints, listProjectOperations, listProjectPermissions, listProjectVpcEndpoints, listProjects, listSharedProjects, listSnapshots, recoverProject, removeOrganizationMember, resetProjectBranchRolePassword, restartProjectEndpoint, restoreProjectBranch, restoreSnapshot, revokeApiKey, revokeOrgApiKey, revokePermissionFromProject, sendNeonAuthTestEmail, setDefaultProjectBranch, setOrganizationSpendingLimit, setSnapshotSchedule, startAnonymization, startProjectEndpoint, suspendProjectEndpoint, transferNeonAuthProviderProject, transferProjectsFromOrgToOrg, transferProjectsFromUserToOrg, updateBranchNeonAuthOauthProvider, updateMaskingRules, updateNeonAuthAllowLocalhost, updateNeonAuthConfig, updateNeonAuthEmailAndPasswordConfig, updateNeonAuthEmailProvider, updateNeonAuthEmailServer, updateNeonAuthMagicLinkPlugin, updateNeonAuthOauthProvider, updateNeonAuthOrganizationPlugin, updateNeonAuthPhoneNumberPlugin, updateNeonAuthUserRole, updateNeonAuthWebhookConfig, updateOrganizationMember, updateProject, updateProjectBranch, updateProjectBranchDataApi, updateProjectBranchDatabase, updateProjectEndpoint, updateSnapshot };
